// @porschehn/navigation/dist/collection/components/menu-desktop-level-1/menu-desktop-level-1.js
import { Component, h } from '@stencil/core';
import { HoverIntentService } from '../../services/hover-intent-service';
import { OverlayAnimation } from '../../services/overlay-animation';
export class MenuDesktopLevel1 {
    constructor() {
        this.hoverIntentService = new HoverIntentService();
        this.overlayAnimation = new OverlayAnimation();
        this.mouseMoveListener = (event) => this.onmousemove(event);
        this.handleMouseLeave = () => {
            document.removeEventListener('mousemove', this.mouseMoveListener);
            this.hoverIntentService.handleHoverOut();
            this.mouseLeftNavigation = true;
        };
        this.showOverlayHandler = (overlayElement) => {
            clearTimeout(this.analyticsCaptureTimeout);
            this.analyticsCaptureTimeout = setTimeout(() => {
                const megaMenuSelected = overlayElement.closest('phn-mega-fly-out-overlay');
                const megaMenuButtonClasses = megaMenuSelected.querySelector('.mega-fly-out-overlay-button').classList;
                if (megaMenuButtonClasses.contains('mega-fly-out-overlay-button-active')) {
                    const openOverlay = megaMenuSelected.querySelector('phn-car-series-model-overview') ||
                        megaMenuSelected.querySelector('phn-shops-and-more-overview');
                    if (openOverlay !== null) {
                        openOverlay.triggerAnalyticsEvent();
                    }
                }
            }, 500);
            if (this.previousOverlay) {
                this.overlayAnimation.show(overlayElement);
            }
            else {
                this.overlayAnimation.slideDown(overlayElement);
            }
        };
        this.hideOverlayHandler = (overlayElement) => {
            if (this.mouseLeftNavigation) {
                this.previousOverlay = undefined;
            }
            else {
                this.previousOverlay = overlayElement;
            }
            if (this.mouseLeftNavigation) {
                this.overlayAnimation.slideUp(overlayElement);
            }
            else {
                this.overlayAnimation.hide(overlayElement);
            }
        };
        this.keyboardShowOverlayHandler = (overlayElement) => {
            this.previousOverlay = overlayElement;
            this.overlayAnimation.slideDown(overlayElement);
            const megaMenuSelected = overlayElement.closest('phn-mega-fly-out-overlay');
            const openOverlay = megaMenuSelected.querySelector('phn-car-series-model-overview') ||
                megaMenuSelected.querySelector('phn-shops-and-more-overview');
            openOverlay.triggerAnalyticsEvent();
        };
        this.keyboardHideOverlayHandler = () => {
            if (this.previousOverlay) {
                this.overlayAnimation.slideUp(this.previousOverlay);
            }
            this.previousOverlay = undefined;
        };
    }
    handleMouseEnter() {
        document.addEventListener('mousemove', this.mouseMoveListener);
        this.mouseLeftNavigation = false;
    }
    onmousemove(event) {
        this.hoverIntentService.handleMouseMove(event, '.overlay-trigger');
    }
    render() {
        return (h("div", { class: "menu-desktop-level-1" },
            h("div", { class: "menu-series-shop-and-more", onMouseEnter: () => this.handleMouseEnter(), onMouseLeave: () => this.handleMouseLeave() },
                h("phn-car-series-list", { showOverlayHandler: this.showOverlayHandler, hideOverlayHandler: this.hideOverlayHandler, keyboardShowOverlayHandler: this.keyboardShowOverlayHandler, keyboardHideOverlayHandler: this.keyboardHideOverlayHandler }),
                h("phn-shops-and-more", { showOverlayHandler: this.showOverlayHandler, hideOverlayHandler: this.hideOverlayHandler, keyboardShowOverlayHandler: this.keyboardShowOverlayHandler, keyboardHideOverlayHandler: this.keyboardHideOverlayHandler })),
            h("div", { class: "menu-separator" }),
            h("phn-meta-functionalities", { class: "metafunctionalities" })));
    }
    static get is() { return "phn-menu-desktop-level-1"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["menu-desktop-level-1.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["menu-desktop-level-1.css"]
    }; }
}
