// @porschehn/navigation/dist/collection/components/menu-desktop-level-1/menu-desktop-level-1.js
import { Component, h, Event } from '@stencil/core';
import { HoverIntentService } from '../../services/hover-intent-service';
import { OverlayAnimation } from '../../services/overlay-animation';
import { navStateStore } from '../../state/nav-state-store';
import { isFeatureEnabled } from '../../env-config';
import { FEATURE_NEW_LEVEL2 } from '../../utility/constants';
export class MenuDesktopLevel1 {
    constructor() {
        this.hoverIntentService = new HoverIntentService();
        this.overlayAnimation = new OverlayAnimation();
        this.mouseMoveListener = (event) => this.onmousemove(event);
        this.handleMouseLeave = () => {
            document.removeEventListener('mousemove', this.mouseMoveListener);
            this.hoverIntentService.handleHoverOut();
            this.mouseLeftNavigation = true;
        };
        this.showOverlayHandler = (overlayElement) => {
            clearTimeout(this.analyticsCaptureTimeout);
            this.analyticsCaptureTimeout = setTimeout(() => {
                const megaMenuSelected = overlayElement.closest('phn-mega-fly-out-overlay');
                const megaMenuButtonActive = megaMenuSelected.querySelector('.mega-fly-out-overlay-button-active');
                if (megaMenuButtonActive !== null) {
                    const openOverlay = megaMenuSelected.querySelector('phn-car-series-model-overview') ||
                        megaMenuSelected.querySelector('phn-shops-and-more-overview');
                    if (openOverlay !== null) {
                        openOverlay.triggerAnalyticsEvent();
                    }
                    else if (isFeatureEnabled(FEATURE_NEW_LEVEL2)) {
                        this.overlayEvent.emit(megaMenuButtonActive.getAttribute('id'));
                    }
                }
            }, 500);
            if (this.previousOverlay) {
                this.overlayAnimation.show(overlayElement);
            }
            else {
                this.overlayAnimation.slideDown(overlayElement);
            }
        };
        this.hideOverlayHandler = (overlayElement) => {
            if (this.mouseLeftNavigation) {
                this.previousOverlay = undefined;
            }
            else {
                this.previousOverlay = overlayElement;
            }
            if (this.mouseLeftNavigation) {
                this.overlayAnimation.slideUp(overlayElement);
            }
            else {
                this.overlayAnimation.hide(overlayElement);
            }
        };
        this.keyboardShowOverlayHandler = (overlayElement) => {
            this.previousOverlay = overlayElement;
            this.overlayAnimation.slideDown(overlayElement);
            const megaMenuSelected = overlayElement.closest('phn-mega-fly-out-overlay');
            const openOverlay = megaMenuSelected.querySelector('phn-car-series-model-overview') ||
                megaMenuSelected.querySelector('phn-shops-and-more-overview');
            if (openOverlay !== null) {
                openOverlay.triggerAnalyticsEvent();
            }
            else if (isFeatureEnabled(FEATURE_NEW_LEVEL2)) {
                this.overlayEvent.emit(megaMenuSelected.querySelector('button').getAttribute('id'));
            }
        };
        this.keyboardHideOverlayHandler = () => {
            if (this.previousOverlay) {
                this.overlayAnimation.slideUp(this.previousOverlay);
            }
            this.previousOverlay = undefined;
        };
    }
    handleMouseEnter() {
        if (navStateStore.get('navigationLoaded')) {
            document.addEventListener('mousemove', this.mouseMoveListener);
            this.mouseLeftNavigation = false;
        }
    }
    onmousemove(event) {
        this.hoverIntentService.handleMouseMove(event, '.overlay-trigger');
    }
    render() {
        return (h("div", { class: "menu-desktop-level-1" },
            h("div", { class: "menu-series-shop-and-more", onMouseEnter: () => this.handleMouseEnter(), onMouseLeave: () => this.handleMouseLeave() },
                h("phn-car-series-list", { showOverlayHandler: this.showOverlayHandler, hideOverlayHandler: this.hideOverlayHandler, keyboardShowOverlayHandler: this.keyboardShowOverlayHandler, keyboardHideOverlayHandler: this.keyboardHideOverlayHandler }),
                h("phn-shops-and-more", { showOverlayHandler: this.showOverlayHandler, hideOverlayHandler: this.hideOverlayHandler, keyboardShowOverlayHandler: this.keyboardShowOverlayHandler, keyboardHideOverlayHandler: this.keyboardHideOverlayHandler })),
            navStateStore.get('navigationLoaded') ? h("div", { class: "menu-separator" }) : null,
            h("phn-meta-functionalities", { class: "metafunctionalities" })));
    }
    static get is() { return "phn-menu-desktop-level-1"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["menu-desktop-level-1.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["menu-desktop-level-1.css"]
    }; }
    static get events() { return [{
            "method": "overlayEvent",
            "name": "overlayEvent",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            }
        }]; }
}
