// @porschehn/navigation/dist/collection/components/menu-desktop-level-1/menu-desktop-level-1.js
import { Component, Prop, h } from '@stencil/core';
import StateTunnel from '../../services/state-tunnel';
import { HoverIntentService } from '../../services/hover-intent-service';
import { OverlayAnimation } from '../../services/overlay-animation';
import { DEFAULT_SERIES_LIST } from '../../services/constants';
export class MenuDesktopLevel1 {
    constructor() {
        this.hoverIntentService = new HoverIntentService();
        this.overlayAnimation = new OverlayAnimation();
        this.mouseMoveListener = (event) => this.onmousemove(event);
        this.handleMouseLeave = () => {
            document.removeEventListener('mousemove', this.mouseMoveListener);
            this.hoverIntentService.handleHoverOut();
            this.mouseLeftNavigation = true;
        };
        this.showOverlayHandler = (overlayElement) => {
            clearTimeout(this.analyticsCaptureTimeout);
            this.analyticsCaptureTimeout = setTimeout(() => {
                const megaMenuSelected = overlayElement.closest('phn-mega-fly-out-overlay');
                const megaMenuButtonClasses = megaMenuSelected.querySelector('.mega-fly-out-overlay-button').classList;
                if (megaMenuButtonClasses.contains('mega-fly-out-overlay-button-active')) {
                    const openOverlay = megaMenuSelected.querySelector('phn-car-series-model-overview') ||
                        megaMenuSelected.querySelector('phn-shops-and-more-overview');
                    if (openOverlay !== null) {
                        openOverlay.triggerAnalyticsEvent();
                    }
                }
            }, 500);
            if (this.previousOverlay) {
                this.overlayAnimation.show(overlayElement);
            }
            else {
                this.overlayAnimation.slideDown(overlayElement);
            }
        };
        this.hideOverlayHandler = (overlayElement) => {
            if (this.mouseLeftNavigation) {
                this.previousOverlay = undefined;
            }
            else {
                this.previousOverlay = overlayElement;
            }
            if (this.mouseLeftNavigation) {
                this.overlayAnimation.slideUp(overlayElement);
            }
            else {
                this.overlayAnimation.hide(overlayElement);
            }
        };
        this.keyboardShowOverlayHandler = (overlayElement) => {
            this.previousOverlay = overlayElement;
            this.overlayAnimation.slideDown(overlayElement);
        };
        this.keyboardHideOverlayHandler = () => {
            if (this.previousOverlay) {
                this.previousOverlay.closest('phn-mega-fly-out-overlay').deactivateState();
                this.overlayAnimation.slideUp(this.previousOverlay);
            }
            this.previousOverlay = undefined;
        };
    }
    handleMouseEnter() {
        document.addEventListener('mousemove', this.mouseMoveListener);
        this.mouseLeftNavigation = false;
    }
    onmousemove(event) {
        this.hoverIntentService.handleMouseMove(event, '.overlay-trigger');
    }
    getItems() {
        const items = [];
        if (this.shops && this.shops.length > 0) {
            items.push(this.shops[0]);
        }
        if (this.more && this.more.length > 0) {
            items.push(this.more[0]);
        }
        return items;
    }
    render() {
        const seriesList = !this.seriesList ? DEFAULT_SERIES_LIST : this.seriesList;
        const items = this.getItems();
        return (h("div", { class: "menu-desktop-level-1" },
            h("div", { class: "menu-series-shop-and-more", onMouseEnter: () => this.handleMouseEnter(), onMouseLeave: () => this.handleMouseLeave() },
                h("phn-car-series-list", { carSeries: seriesList, showOverlayHandler: this.showOverlayHandler, hideOverlayHandler: this.hideOverlayHandler, keyboardShowOverlayHandler: this.keyboardShowOverlayHandler, keyboardHideOverlayHandler: this.keyboardHideOverlayHandler }),
                h("phn-shops-and-more", { items: items, showOverlayHandler: this.showOverlayHandler, hideOverlayHandler: this.hideOverlayHandler, keyboardShowOverlayHandler: this.keyboardShowOverlayHandler, keyboardHideOverlayHandler: this.keyboardHideOverlayHandler })),
            h("div", { class: "menu-separator" }),
            h("phn-meta-functionalities", { class: "metafunctionalities" })));
    }
    static get is() { return "phn-menu-desktop-level-1"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["menu-desktop-level-1.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["menu-desktop-level-1.css"]
    }; }
    static get properties() { return {
        "seriesList": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Series[]",
                "resolved": "Series[]",
                "references": {
                    "Series": {
                        "location": "import",
                        "path": "../entities/series"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "shops": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "TextItem[]",
                "resolved": "TextItem[]",
                "references": {
                    "TextItem": {
                        "location": "import",
                        "path": "../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "more": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "TextItem[]",
                "resolved": "TextItem[]",
                "references": {
                    "TextItem": {
                        "location": "import",
                        "path": "../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
}
StateTunnel.injectProps(MenuDesktopLevel1, ['shops', 'more', 'seriesList']);
