// @porschehn/navigation/dist/collection/components/car-model-list-v2/car-model-list-v2.js
import { Component, Prop, h } from '@stencil/core';
import { ModelClickEvent } from '../../services/datalayer/model-click-event';
import { EMPTY_SERIES } from '../../utility/constants';
import { analyticsSetupStore } from '../../state/analytics-setup-store';
export class CarModelListV2 {
    constructor() {
        this.series = EMPTY_SERIES;
    }
    sendModelClickAnalyticsEvent(model) {
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window'),
            environment: analyticsSetupStore.get('environment')
        };
        const modelClickEvent = new ModelClickEvent(analyticsSetup, this.series.Name, model);
        const pagData = (window['pagData'] = window['pagData'] || []);
        pagData.push(modelClickEvent.event());
    }
    render() {
        return (h("div", { class: "car-model-list" }, this.series.Models.map((model) => (h("phn-p-link", { variant: "tertiary", href: model.Link, icon: 'arrow-head-right', size: "inherit", class: "car-model-list-item", onClick: () => this.sendModelClickAnalyticsEvent(model) }, model.Name)))));
    }
    static get is() { return "phn-car-model-list-v2"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["car-model-list-v2.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["car-model-list-v2.css"]
    }; }
    static get properties() { return {
        "series": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Series",
                "resolved": "Series",
                "references": {
                    "Series": {
                        "location": "import",
                        "path": "../../entities/series"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_SERIES"
        }
    }; }
}
