// @porschehn/navigation/dist/collection/components/car-meta-functionalities/car-meta-functionalities.js
import { Component, Prop, h } from '@stencil/core';
import { ConfigureClickEvent } from '../../services/datalayer/configure-click-event';
import { CompareClickEvent } from '../../services/datalayer/compare-click-event';
import { FindCarClickEvent } from '../../services/datalayer/find-car-click-event';
import { analyticsSetupStore } from '../../state/analytics-setup-store';
export class CarMetaFunctionalities {
    constructor() {
        this.items = [
            { PhraseId: '', Text: '', Link: '' },
            { PhraseId: '', Text: '', Link: '' },
            { PhraseId: '', Text: '', Link: '' }
        ];
        this.sendConfigureClickAnalyticsEvent = (contentItem) => {
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const configureClickEvent = new ConfigureClickEvent(analyticsSetup, this.name, contentItem);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(configureClickEvent.event());
        };
        this.sendCompareClickAnalyticsEvent = (contentItem) => {
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const compareClickEvent = new CompareClickEvent(analyticsSetup, this.name, contentItem);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(compareClickEvent.event());
        };
        this.sendFindCarClickAnalyticsEvent = (contentItem) => {
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const findCarClickEvent = new FindCarClickEvent(analyticsSetup, this.name, contentItem);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(findCarClickEvent.event());
        };
    }
    renderItem(source, contentItem, analyticsEvent) {
        return (h("p-link-pure", { href: contentItem.Link, icon: source, size: "inherit", onClick: () => analyticsEvent(contentItem), class: "car-meta-functionalities-item" }, contentItem.Text));
    }
    render() {
        return (h("div", { class: "car-meta-functionalities" },
            h("div", { class: "car-meta-functionalities-wrapper" },
                this.renderItem('configurate', this.items[0], this.sendConfigureClickAnalyticsEvent),
                this.renderItem('compare', this.items[1], this.sendCompareClickAnalyticsEvent),
                this.renderItem('search', this.items[2], this.sendFindCarClickAnalyticsEvent))));
    }
    static get is() { return "phn-car-meta-functionalities"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["car-meta-functionalities.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["car-meta-functionalities.css"]
    }; }
    static get properties() { return {
        "name": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "name",
            "reflect": false
        },
        "items": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ContentItem[]",
                "resolved": "ContentItem[]",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "[\n    { PhraseId: '', Text: '', Link: '' },\n    { PhraseId: '', Text: '', Link: '' },\n    { PhraseId: '', Text: '', Link: '' }\n  ]"
        }
    }; }
}
