// @porschehn/navigation/dist/collection/components/refactoring/mobile-car-series-list/mobile-car-series-list.js
import { Component, h, Prop } from '@stencil/core';
import { DEFAULT_SERIES_LIST, FEATURE_KEYBOARD } from '../../../services/constants';
import { navStateStore } from '../../../services/nav-state-store';
import { isFeatureEnabled } from '../../../env-config';
import { navContentStore } from '../../../services/nav-content-store';
import { analyticsSetupStore } from '../../../services/datalayer/analytics-setup-store';
import { ModelRangeClickEvent } from '../../../services/datalayer/model-range-click-event';
export class MobileCarSeriesList {
    constructor() {
        this.carSeries = DEFAULT_SERIES_LIST;
        this.keyDownHandler = (event, series, index) => {
            if (isFeatureEnabled(FEATURE_KEYBOARD, navContentStore.get('env'))) {
                if (event.key === 'Enter') {
                    this.showDrawer(event, series, index);
                }
            }
        };
        this.sendNavigationClickAnalyticsEvent = (series) => {
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const modelSeriesClickEvent = new ModelRangeClickEvent(analyticsSetup, series.Name, series);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(modelSeriesClickEvent.event());
        };
    }
    showDrawer(_, series, index) {
        const phnDrawerCarSeriesModelOverview2 = (h("phn-drawer-car-series-model-overview-2", { series: series, class: 'drawer-car-series-model-overview drawer-car-series-model-overview-' + index, slot: "content" }));
        const parent = navStateStore.get('activeDrawerContent');
        const currentParent = navStateStore.get('parentDrawerContent');
        navStateStore.set('activeDrawerContent', phnDrawerCarSeriesModelOverview2);
        navStateStore.set('parentDrawerContent', [parent, ...currentParent]);
        navStateStore.set('accessibilityMarker', `.car-series-${index}`);
        navStateStore.set('backButtonText', null);
        this.sendNavigationClickAnalyticsEvent(series);
    }
    render() {
        return (h("div", { class: "mobile-car-series-list" }, this.carSeries.map((series, index) => (h("div", { class: "mobile-car-series" },
            h("phn-car-series", { class: 'car-series car-series-' + index, text: series.Name, image: series.Image, tabindex: "0", onClick: (event) => this.showDrawer(event, series, index), onKeyDown: (event) => this.keyDownHandler(event, series, index) }))))));
    }
    static get is() { return "phn-mobile-car-series-list-2"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["mobile-car-series-list.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["mobile-car-series-list.css"]
    }; }
    static get properties() { return {
        "carSeries": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Series[]",
                "resolved": "Series[]",
                "references": {
                    "Series": {
                        "location": "import",
                        "path": "../../entities/series"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "DEFAULT_SERIES_LIST"
        }
    }; }
}
