// @porschehn/navigation/dist/collection/state/nav-state-store.js
import { createStore } from '@stencil/store';
export const navStateStore = createStore({
    drawerContentStack: [],
    isDrawerOpen: false,
    imagesRegistered: 0,
    imagesRetrieved: 0,
    imagesLoaded: true,
    navigationLoaded: false,
    crestLoaded: false,
    pinSvgLoaded: false,
    userSvgLoaded: false,
    spinnerSvgLoaded: false,
    loggedInUserSvgLoaded: false
});
navStateStore.onChange('imagesRetrieved', () => {
    navStateStore.set('imagesLoaded', navStateStore.get('imagesRetrieved') === navStateStore.get('imagesRegistered'));
});
navStateStore.onChange('imagesRegistered', () => {
    navStateStore.set('imagesLoaded', navStateStore.get('imagesRetrieved') === navStateStore.get('imagesRegistered'));
});
navStateStore.onChange('imagesLoaded', () => {
    navStateStore.set('navigationLoaded', navStateStore.get('crestLoaded') && navStateStore.get('imagesLoaded'));
});
navStateStore.onChange('crestLoaded', () => {
    navStateStore.set('navigationLoaded', navStateStore.get('crestLoaded') && navStateStore.get('imagesLoaded'));
});
