// @porschehn/navigation/dist/collection/utility/constants.js
export const DEFAULT_LOCALE = 'de-CH';
export const DEFAULT_ITEMS_LIST = [];
export const DEFAULT_SERIES_LIST = [];
export const DEFAULT_SHOPS = [];
export const DEFAULT_MORE = [];
export const DEFAULT_METAFUNCTIONALITIES = [];
export const DEFAULT_MENU = [];
export const DEFAULT_ENVIRONMENT = 'ENVIRONMENT';
export const ANALYTICS_APP_ID = 'nav';
export const EMPTY_SERIES = { PhraseId: '', Name: '', Image: '', Models: [], Further: [], Meta: [] };
export const EMPTY_ITEM = { PhraseId: '', Text: '' };
export const REDACTED = 'REDACTED';
export const LOGGED_IN_USER_NO_NAME_STATUS = 'loggedIn';
export const LOGGED_OUT_USER_STATUS = 'loginOrRegister';
export const LOADING_STATUS = 'loading';
export const VERSION = 'VERSION_STRING';
export const DEFAULT_UNREAD_MESSAGES_COUNT = 0;
export const DEFAULT_EMPTY_CONTENT_RESPONSE = [];
export const DEFAULT_EMPTY_CONTENT = {
    seriesList: DEFAULT_SERIES_LIST,
    shops: DEFAULT_SHOPS,
    more: DEFAULT_MORE,
    metafunctionalities: DEFAULT_METAFUNCTIONALITIES,
    menu: DEFAULT_MENU
};
export const ACCESS_TOKEN_FETCHED_EVENT = 'accessTokenFetched';
export const ACCESS_TOKEN_FETCH_ERROR_EVENT = 'accessTokenFetchError';
export const SLOT_UPDATED_EVENT = 'slotUpdated';
export const TEXT_FIELD_NAME = 'Text';
export const LINK_FIELD_NAME = 'Link';
export const PHRASEID_FIELD_NAME = 'PhraseId';
export const DEALER_SEARCH_PHRASE_ID = 'metafunctionalities.finddealer';
export const MYPORSCHE_PHRASE_ID = 'metafunctionalities.myporsche';
export const CREST_PHRASE_ID = 'mainmenu.crest';
export const MENU_PHRASE_ID = 'mainmenu.menu';
export const FEATURE_OVERRIDE_ENABLED = 'featureOverrideEnabled';
// Defining the feature names
export const FEATURE_FEATURE1 = 'homenav_feature1';
export const MAX_M_VIEWPORT_WIDTH = 1299;
export const FOCUSABLE_ELEMENTS_SELECTOR = 'a, button, [tabindex]:not([tabindex="-1"]), p-link-pure, p-button-pure';
