// @porschehn/navigation/dist/collection/components/mobile-shops-and-more/mobile-shops-and-more.js
import { Component, Host, h } from '@stencil/core';
import { navContentStore } from '../../state/nav-content-store';
import { navStateStore } from '../../state/nav-state-store';
import { analyticsSetupStore } from '../../state/analytics-setup-store';
import { NavigationClickEvent } from '../../services/datalayer/navigation-click-event';
export class MobileShopsAndMore {
    constructor() {
        this.showDrawer = (_, item) => {
            const phnDrawerShopsAndMore = {
                content: (h("phn-drawer-shops-and-more", { headline: item.Text, item: item.Items && item.Items.length === 1 ? item.Items[0] : item, slot: "content", class: "shops-and-more-drawer-level-1" }))
            };
            navStateStore.set('drawerContentStack', [...navStateStore.get('drawerContentStack'), phnDrawerShopsAndMore]);
            this.sendNavigationClickAnalyticsEvent(item);
        };
        this.handleKeyDown = (event, item) => {
            if (event.key === 'Enter') {
                this.showDrawer(event, item);
            }
        };
        this.sendNavigationClickAnalyticsEvent = (item) => {
            const contentItem = { PhraseId: item.PhraseId, Text: item.Text, Link: undefined };
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const navigationClickEvent = new NavigationClickEvent(analyticsSetup, contentItem);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(navigationClickEvent.event());
        };
    }
    getItems() {
        const items = [];
        const shops = navContentStore.get('shops');
        const more = navContentStore.get('more');
        if (shops && shops.length > 0) {
            items.push(shops[0]);
        }
        if (more && more.length > 0) {
            items.push(more[0]);
        }
        return items;
    }
    render() {
        const items = this.getItems();
        return (h(Host, { class: "mobile-shops-and-more" }, items.map((item, index) => {
            return (h("div", { class: "shops-and-more-container" },
                h("phn-mobile-shop-and-more-item", { item: item, class: 'mobile-shop-and-more-item mobile-shop-and-more-item-' + index, tabindex: "0", onClick: (event) => this.showDrawer(event, item), onKeyDown: (event) => this.handleKeyDown(event, item) })));
        })));
    }
    static get is() { return "phn-mobile-shops-and-more"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["mobile-shops-and-more.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["mobile-shops-and-more.css"]
    }; }
}
