// @porschehn/navigation/dist/collection/components/myporsche-items/test/mock-modules/state-tunnel-mock.js
/* istanbul ignore file */
import { h } from '@stencil/core';
import { createProviderConsumer } from '@stencil/state-tunnel';
import { DEFAULT_ENVIRONMENT } from '../../../../services/constants';
export const messages = {
    Id: 'metafunctionalities.myporsche.messages',
    Text: 'messages',
    Link: 'link'
};
export const carsAndServices = {
    Id: 'metafunctionalities.myporsche.carsandservices',
    Text: 'car and services',
    Link: 'car and services link'
};
export const accountAdministration = {
    Id: 'metafunctionalities.myporsche.accountadministration',
    Text: 'account administration',
    Link: 'account administration link'
};
export const logout = {
    Id: 'metafunctionalities.myporsche.logout',
    Text: 'logout text',
    Link: 'logout link'
};
const metafunctionalities = [
    {
        PhraseId: 'metafunctionalities.myporsche',
        Text: 'my porsche',
        Link: 'link',
        Items: [
            {
                Id: 'metafunctionalities.myporsche.loggedIn',
                Text: 'Logged in user'
            },
            messages,
            carsAndServices,
            accountAdministration,
            logout
        ]
    }
];
export default createProviderConsumer({
    seriesList: [],
    shops: [],
    more: [],
    metafunctionalities,
    menu: [],
    env: DEFAULT_ENVIRONMENT,
    loggedInState: undefined
}, (subscribe, child) => h("context-consumer", { subscribe: subscribe, renderer: child }));
