// @porsche-design-system/components-js/dist/collection/components/form/radio-button-wrapper/radio-button-wrapper.js
import { Host, Component, Prop, h, Element, State } from '@stencil/core';
import cx from 'classnames';
import { mapBreakpointPropToPrefixedClasses, prefix, transitionListener, insertSlottedStyles, randomString } from '../../../utils';
export class RadioButtonWrapper {
    constructor() {
        /** The label text. */
        this.label = '';
        /** The validation state. */
        this.state = 'none';
        /** The message styled depending on validation state. */
        this.message = '';
        /** Show or hide label. For better accessibility it's recommended to show the label. */
        this.hideLabel = false;
        this.labelId = randomString();
    }
    componentWillLoad() {
        this.setInput();
        this.setState();
        this.bindStateListener();
        this.addSlottedStyles();
    }
    render() {
        const labelClasses = cx(prefix('radio-button-wrapper__label'));
        const fakeRadioButtonClasses = cx(prefix('radio-button-wrapper__fake-radio-button'), this.checked && prefix(`radio-button-wrapper__fake-radio-button--checked`), this.disabled && prefix(`radio-button-wrapper__fake-radio-button--disabled`), this.state !== 'none' && prefix(`radio-button-wrapper__fake-radio-button--${this.state}`));
        const labelTextClasses = cx(prefix('radio-button-wrapper__label-text'), mapBreakpointPropToPrefixedClasses('radio-button-wrapper__label-text-', this.hideLabel, ['hidden', 'visible']), this.disabled && prefix('radio-button-wrapper__label-text--disabled'));
        const messageClasses = cx(prefix('radio-button-wrapper__message'), this.state !== 'none' && prefix(`radio-button-wrapper__message--${this.state}`));
        return (h(Host, null,
            h("label", { class: labelClasses, id: this.state === 'error' && this.labelId },
                this.isLabelVisible &&
                    h("p-text", { class: labelTextClasses, tag: 'span', color: 'inherit', onClick: (e) => this.labelClick(e) }, this.label ? this.label : h("span", null,
                        h("slot", { name: 'label' }))),
                h("span", { class: fakeRadioButtonClasses },
                    h("slot", null))),
            this.isMessageVisible &&
                h("p-text", { class: messageClasses, color: 'inherit', role: this.state === 'error' && 'alert', "aria-describedby": this.state === 'error' && this.labelId }, this.message ? this.message : h("span", null,
                    h("slot", { name: 'message' })))));
    }
    get isLabelVisible() {
        return !!this.label || !!this.host.querySelector('[slot="label"]');
    }
    get isMessageDefined() {
        return !!this.message || !!this.host.querySelector('[slot="message"]');
    }
    get isMessageVisible() {
        return ['success', 'error'].includes(this.state) && this.isMessageDefined;
    }
    setInput() {
        this.input = this.host.querySelector('input[type="radio"]');
        this.input.setAttribute('aria-label', this.label);
    }
    labelClick(event) {
        /**
         * we only want to simulate the checkbox click by label click
         * for real shadow dom, else the native behaviour works out
         * of the box
         */
        if (this.host.shadowRoot && this.host.shadowRoot.host
            && event.target.closest('a') === null) {
            this.input.click();
        }
    }
    setState() {
        this.checked = this.input.checked;
        this.disabled = this.input.disabled;
    }
    bindStateListener() {
        transitionListener(this.input, 'border-top-color', () => {
            this.setState();
        });
    }
    addSlottedStyles() {
        const tagName = this.host.tagName.toLowerCase();
        const style = `${tagName} a {
      outline: none transparent;
      color: inherit;
      text-decoration: underline;
      -webkit-transition: outline-color .24s ease, color .24s ease;
      transition: outline-color .24s ease, color .24s ease;
    }

    ${tagName} a:hover {
      color: #d5001c;
    }

    ${tagName} a:focus {
      outline: 2px solid #00d5b9;
      outline-offset: 1px;
    }`;
        insertSlottedStyles(this.host, style);
    }
    static get is() { return "p-radio-button-wrapper"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["radio-button-wrapper.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["radio-button-wrapper.css"]
    }; }
    static get properties() { return {
        "label": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "The label text."
            },
            "attribute": "label",
            "reflect": false,
            "defaultValue": "''"
        },
        "state": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "FormState",
                "resolved": "\"error\" | \"none\" | \"success\"",
                "references": {
                    "FormState": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "The validation state."
            },
            "attribute": "state",
            "reflect": false,
            "defaultValue": "'none'"
        },
        "message": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "The message styled depending on validation state."
            },
            "attribute": "message",
            "reflect": false,
            "defaultValue": "''"
        },
        "hideLabel": {
            "type": "any",
            "mutable": false,
            "complexType": {
                "original": "BreakpointCustomizable<boolean>",
                "resolved": "BreakpointValues<boolean> | boolean | string",
                "references": {
                    "BreakpointCustomizable": {
                        "location": "import",
                        "path": "../../../utils"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Show or hide label. For better accessibility it's recommended to show the label."
            },
            "attribute": "hide-label",
            "reflect": false,
            "defaultValue": "false"
        }
    }; }
    static get states() { return {
        "checked": {},
        "disabled": {}
    }; }
    static get elementRef() { return "host"; }
}
