// @porschehn/navigation/dist/collection/components/car-model-list/car-model-list.js
import { Component, Prop, h } from '@stencil/core';
import { ModelClickEvent } from '../../services/datalayer/model-click-event';
import { EMPTY_SERIES } from '../../services/constants';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
export class CarModelList {
    constructor() {
        this.series = EMPTY_SERIES;
    }
    sendModelClickAnalyticsEvent(model) {
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window'),
            environment: analyticsSetupStore.get('environment')
        };
        const modelClickEvent = new ModelClickEvent(analyticsSetup, this.series.Name, model);
        const pagData = (window['pagData'] = window['pagData'] || []);
        pagData.push(modelClickEvent.event());
    }
    render() {
        return (h("div", { class: "car-model-list" },
            h("div", { class: 'car-model-list-wrapper car-models-count-' + this.series.Models.length }, this.series.Models.map((model, index) => (h("div", { class: 'car-model car-model-' + index },
                h("phn-car-model", { text: model.Name, image: model.Image, link: model.Link, onClick: () => this.sendModelClickAnalyticsEvent(model) })))))));
    }
    static get is() { return "phn-car-model-list"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["car-model-list.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["car-model-list.css"]
    }; }
    static get properties() { return {
        "series": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Series",
                "resolved": "Series",
                "references": {
                    "Series": {
                        "location": "import",
                        "path": "../entities/series"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_SERIES"
        }
    }; }
}
