// @porschehn/navigation/dist/collection/components/messages-link/messages-link.js
import { Component, Host, h, Prop } from '@stencil/core';
import { analyticsSetupStore } from '../../state/analytics-setup-store';
import { navContentStore } from '../../state/nav-content-store';
import { EMPTY_ITEM } from '../../utility/constants';
import { NavigationClickEvent } from '../../services/datalayer/navigation-click-event';
export class MessagesLink {
    constructor() {
        this.item = EMPTY_ITEM;
        this.sendNavigationClickAnalyticsEvent = () => {
            const contentItem = { PhraseId: this.item.PhraseId, Text: this.item.Text, Link: this.item.Link };
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const navigationClick = new NavigationClickEvent(analyticsSetup, contentItem);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(navigationClick.event());
        };
    }
    render() {
        const unreadMessagesCount = navContentStore.get('unreadMessagesCount');
        return (h(Host, null,
            h("phn-p-text", { class: "list-item-link", tag: "div", color: "inherit", size: "inherit" },
                h("a", { href: this.item.Link, onClick: this.sendNavigationClickAnalyticsEvent },
                    h("span", { class: "link-text" }, this.item.Text),
                    h("span", { class: unreadMessagesCount > 0 ? 'unread-messages-count' : 'unread-messages-count hidden' }, unreadMessagesCount)))));
    }
    static get is() { return "phn-messages-link"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["messages-link.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["messages-link.css"]
    }; }
    static get properties() { return {
        "item": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ContentItem",
                "resolved": "ContentItem",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_ITEM"
        }
    }; }
}
