// @porschehn/navigation/dist/collection/services/datalayer/model-analytics-event.js
import { AnalyticsEvent } from './analytics-event';
import { CarCollectionDimensions, ComponentClickCollectionDimension } from './custom-dimensions';
export class ModelAnalyticsEvent extends AnalyticsEvent {
    constructor(analyticsSetup, modelRangeName) {
        super(analyticsSetup);
        this.context.eventAction = 'PAGHomeNav_ModelSeries_Click';
        this.initCar();
        this.initComponentClick();
        this.populateCar(modelRangeName);
    }
    initCar() {
        this.car = {
            [CarCollectionDimensions.ModelRangeCode]: undefined,
            [CarCollectionDimensions.ModelRangeName]: undefined,
            [CarCollectionDimensions.ModelSeriesName]: undefined,
            [CarCollectionDimensions.ModelSeriesCode]: undefined,
            [CarCollectionDimensions.ModelName]: undefined,
            [CarCollectionDimensions.Bodytype]: undefined
        };
    }
    initComponentClick() {
        this.componentClick = {
            [ComponentClickCollectionDimension.ClickElementId]: undefined,
            [ComponentClickCollectionDimension.ClickElementName]: undefined,
            [ComponentClickCollectionDimension.ClickElementType]: undefined,
            [ComponentClickCollectionDimension.TargetUrl]: undefined
        };
    }
    populateCar(modelRangeName) {
        this.car.modelRangeName = modelRangeName;
    }
    event() {
        const car = this.car;
        const componentClick = this.componentClick;
        return Object.assign(Object.assign({}, super.event()), { car,
            componentClick });
    }
}
