// @porschehn/navigation/dist/collection/components/drawer-car-series-model-overview/drawer-car-series-model-overview.js
import { Component, Element, Prop, h, Listen } from '@stencil/core';
import { EMPTY_SERIES, MENU_PHRASE_ID, PHRASEID_FIELD_NAME, TEXT_FIELD_NAME, FEATURE_KEYBOARD, FOCUSABLE_ELEMENTS_SELECTOR } from '../../services/constants';
import { isFeatureEnabled } from '../../env-config';
import { navContentStore } from '../../services/nav-content-store';
export class DrawerCarSeriesModelOverview {
    constructor() {
        this.series = EMPTY_SERIES;
        this.handleClick = () => {
            this.el.style.display = 'none';
            this.closeActiveDrawer(2);
        };
    }
    getValueForField(fieldName) {
        let text = '';
        const menu = navContentStore.get('menu');
        if (menu && menu.length > 0) {
            const menuItem = menu.find((m) => m.PhraseId === MENU_PHRASE_ID);
            if (menuItem && menuItem[fieldName]) {
                text = menuItem[fieldName];
            }
        }
        return text;
    }
    keyDownHandler(event) {
        if (isFeatureEnabled(FEATURE_KEYBOARD, navContentStore.get('env'))) {
            if (!event.shiftKey && event.key === 'Tab') {
                const focusable = this.el.querySelectorAll(FOCUSABLE_ELEMENTS_SELECTOR);
                if (event.target === focusable[focusable.length - 1]) {
                    event.preventDefault();
                    this.el.querySelector('phn-back-button > p-button-pure').focus();
                }
            }
        }
    }
    render() {
        const backButtonText = this.getValueForField(TEXT_FIELD_NAME);
        const backButtonId = this.getValueForField(PHRASEID_FIELD_NAME);
        return (h("div", { class: "drawer-car-series-model-overview" },
            h("div", { class: "drawer-container drawer-level-2-item-scroll" },
                h("div", { class: "drawer-back-and-close" },
                    h("phn-back-button", { class: "back-button", onClick: this.handleClick, buttonId: backButtonId, text: backButtonText }),
                    h("phn-close-button", { class: "close-button" })),
                h("phn-headline", { text: this.series.Name }),
                h("phn-car-meta-functionalities", { name: this.series.Name, items: this.series.Meta }),
                h("phn-car-model-list", { class: "car-model-list", series: this.series }),
                h("phn-car-further-links", { items: this.series.Further }))));
    }
    static get is() { return "phn-drawer-car-series-model-overview"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["drawer-car-series-model-overview.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["drawer-car-series-model-overview.css"]
    }; }
    static get properties() { return {
        "series": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Series",
                "resolved": "Series",
                "references": {
                    "Series": {
                        "location": "import",
                        "path": "../entities/series"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_SERIES"
        },
        "closeActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(level: number) => void",
                "resolved": "(level: number) => void",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
    static get elementRef() { return "el"; }
    static get listeners() { return [{
            "name": "keydown",
            "method": "keyDownHandler",
            "target": undefined,
            "capture": false,
            "passive": false
        }]; }
}
