// @porschehn/navigation/dist/collection/components/shops-and-more-item/shops-and-more-item.js
import { Component, Host, Prop, h } from '@stencil/core';
import { NavigationClickEvent } from '../../services/datalayer/navigation-click-event';
import { EMPTY_ITEM } from '../../services/constants';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
export class ShopsAndMoreItem {
    constructor() {
        this.item = EMPTY_ITEM;
    }
    sendNavigationClickAnalyticsEvent(item) {
        const contentItem = { PhraseId: item.PhraseId, Text: item.Text, Link: undefined };
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window')
        };
        const navigationClickEvent = new NavigationClickEvent(analyticsSetup, contentItem);
        const pagData = (window['pagData'] = window['pagData'] || []);
        pagData.push(navigationClickEvent.event());
    }
    render() {
        return (h(Host, null,
            h("phn-button-extended", { class: "shops-and-more-item-drawer", icon: "arrow-head-right", onClick: () => this.sendNavigationClickAnalyticsEvent(this.item) }, this.item.Text),
            h("phn-button-extended", { class: "shops-and-more-item-mega-menu" }, this.item.Text)));
    }
    static get is() { return "phn-shops-and-more-item"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["shops-and-more-item.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["shops-and-more-item.css"]
    }; }
    static get properties() { return {
        "item": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Item",
                "resolved": "Item",
                "references": {
                    "Item": {
                        "location": "import",
                        "path": "../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_ITEM"
        }
    }; }
}
