// @porsche-design-system/components-js/dist/cjs/p-text-list-item.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f34d78e1.js');
const index$1 = require('./index-8af06bf3.js');

const textListItemCss = ":host{position:relative !important;display:list-item !important;color:inherit !important;list-style-type:none !important}:host(.p-text-list-item--unordered){padding-left:1.5rem !important}:host(.p-text-list-item--unordered)::before{content:\"\" !important;position:absolute !important;left:0 !important;top:calc(1.5em / 2 - 0.125em) !important;width:0.25rem !important;height:0.25rem !important;background-color:currentColor !important}:host(.p-text-list-item--unordered.p-text-list-item--nested)::before{height:1px !important;width:0.5rem !important;top:calc(1.5em / 2) !important}:host(.p-text-list-item--ordered){padding-left:2.5rem !important}:host(.p-text-list-item--ordered)::before{position:absolute !important;right:calc(100% - 24px) !important;top:0 !important;counter-increment:section !important;text-align:right !important;font-size:1rem;line-height:1.5;font-family:\"Porsche Next\", \"Arial Narrow\", Arial, sans-serif;font-weight:400}:host(.p-text-list-item--ordered-numbered)::before{content:counters(section, \".\", decimal) \".\" !important}:host(.p-text-list-item--ordered-alphabetically)::before{content:counters(section, \".\", lower-latin) \".\" !important}:host(.p-text-list-item--ordered-numbered.p-text-list-item--nested)::before{content:counters(section, \".\", decimal) !important}:host(.p-text-list-item--ordered-alphabetically.p-text-list-item--nested)::before{content:counters(section, \".\", lower-latin) !important}";

const TextListItem = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
    }
    componentDidLoad() {
        this.addSlottedStyles();
    }
    render() {
        const textListItemClasses = index$1.cx(index$1.prefix('text-list-item'), index$1.prefix(`text-list-item--${this.typeOfList}`), {
            [index$1.prefix(`text-list-item--ordered-${this.typeOfOrderedList}`)]: this.typeOfList === 'ordered',
            [index$1.prefix('text-list-item--nested')]: this.isNestedList
        });
        return (index.h(index.Host, { role: 'listitem', class: textListItemClasses }, index.h("slot", null)));
    }
    get typeOfList() {
        const list = this.host.closest(index$1.prefix('text-list'));
        return list.getAttribute('list-type');
    }
    get typeOfOrderedList() {
        const list = this.host.closest(index$1.prefix('text-list'));
        return list.getAttribute('order-type');
    }
    get isNestedList() {
        return !!this.host.closest(index$1.prefix('text-list[nested]'));
    }
    addSlottedStyles() {
        const tagName = this.host.tagName.toLowerCase();
        const style = `${tagName} a {
      outline: none transparent;
      color: inherit;
      text-decoration: underline;
      -webkit-transition: outline-color .24s ease, color .24s ease;
      transition: outline-color .24s ease, color .24s ease;
    }

    ${tagName} a:hover {
      color: #d5001c;
    }

    ${tagName} a:focus {
      outline: 2px solid #00d5b9;
      outline-offset: 1px;
    }
    `;
        index$1.insertSlottedStyles(this.host, style);
    }
    get host() { return index.getElement(this); }
};
TextListItem.style = textListItemCss;

exports.p_text_list_item = TextListItem;
