// @porschehn/navigation/dist/collection/components/dealer-search/dealer-search.js
import { Component, Host, Prop, h, Element } from '@stencil/core';
import { DEALER_SEARCH_PHRASE_ID } from '../../utility/constants';
import { navStateStore } from '../../state/nav-state-store';
export class DealerSearch {
    constructor() {
        this.dealerSearchText = '';
        this.dealerSearchLink = '';
    }
    render() {
        const hiddenStyle = { display: 'none' };
        return (h(Host, { class: "dealer-search" },
            h("phn-p-link-pure", { style: navStateStore.get('navigationLoaded') ? {} : hiddenStyle, href: this.dealerSearchLink, size: "inherit", icon: "pin", class: 'dealer-search-link', onClick: () => {
                    const linkItem = {
                        PhraseId: DEALER_SEARCH_PHRASE_ID,
                        Text: this.dealerSearchText,
                        Link: this.dealerSearchLink
                    };
                    this.sendMetaFunctionClickAnalyticsEvent(linkItem);
                } }, this.dealerSearchText),
            h("div", { class: "dealer-search-skeleton", style: navStateStore.get('navigationLoaded') ? hiddenStyle : {} })));
    }
    static get is() { return "phn-dealer-search"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["dealer-search.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["dealer-search.css"]
    }; }
    static get properties() { return {
        "dealerSearchText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "dealer-search-text",
            "reflect": false,
            "defaultValue": "''"
        },
        "dealerSearchLink": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "dealer-search-link",
            "reflect": false,
            "defaultValue": "''"
        },
        "sendMetaFunctionClickAnalyticsEvent": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(item: ContentItem) => void",
                "resolved": "(item: ContentItem) => void",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
    static get elementRef() { return "el"; }
}
