// @porschehn/navigation/dist/collection/components/dealer-search/dealer-search.js
import { Component, Host, Prop, h } from '@stencil/core';
import { DEALER_SEARCH_PHRASE_ID } from '../../services/constants';
export class DealerSearch {
    constructor() {
        this.dealerSearchText = '';
        this.dealerSearchLink = '';
    }
    render() {
        return (h(Host, null,
            h("p-link-pure", { href: this.dealerSearchLink, size: "inherit", icon: "pin", onClick: () => {
                    const linkItem = {
                        PhraseId: DEALER_SEARCH_PHRASE_ID,
                        Text: this.dealerSearchText,
                        Link: this.dealerSearchLink
                    };
                    this.sendMetaFunctionClickAnalyticsEvent(linkItem);
                } }, this.dealerSearchText)));
    }
    static get is() { return "phn-dealer-search"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["dealer-search.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["dealer-search.css"]
    }; }
    static get properties() { return {
        "dealerSearchText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "dealer-search-text",
            "reflect": false,
            "defaultValue": "''"
        },
        "dealerSearchLink": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "dealer-search-link",
            "reflect": false,
            "defaultValue": "''"
        },
        "sendMetaFunctionClickAnalyticsEvent": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(item: ContentItem) => void",
                "resolved": "(item: ContentItem) => void",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
}
