// @porschehn/navigation/dist/collection/components/meta-functionalities/meta-functionalities.js
import { Component, Element, Prop, h } from '@stencil/core';
import { MetaFunctionClick } from '../../services/datalayer/meta-function-click';
import { DEALER_SEARCH_PHRASE_ID, LINK_FIELD_NAME, MYPORSCHE_PHRASE_ID, TEXT_FIELD_NAME } from '../../services/constants';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
import { navContentStore } from '../../services/nav-content-store';
export class MetaFunctionalities {
    constructor() {
        this.sendMetaFunctionClickAnalyticsEvent = (linkItem) => {
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const metaFunctionClick = new MetaFunctionClick(analyticsSetup, linkItem);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(metaFunctionClick.event());
        };
    }
    getFieldValueForPhraseId(phraseId, fieldName) {
        const functionality = navContentStore.get('metafunctionalities').find((meta) => meta.PhraseId === phraseId);
        if (!functionality) {
            return '';
        }
        return functionality[fieldName] ? functionality[fieldName] : '';
    }
    getLoggedInStatusTranslations() {
        const functionality = navContentStore
            .get('metafunctionalities')
            .find((meta) => meta.PhraseId === MYPORSCHE_PHRASE_ID);
        const loggedInStatusTranslations = {};
        if (functionality && functionality.Items) {
            functionality.Items.forEach((meta) => {
                loggedInStatusTranslations[meta.PhraseId] = meta[TEXT_FIELD_NAME];
            });
        }
        return loggedInStatusTranslations;
    }
    render() {
        return (h("div", { class: "metafunctionalities" },
            h("phn-dealer-search", { class: "dealer-search", dealerSearchText: this.getFieldValueForPhraseId(DEALER_SEARCH_PHRASE_ID, TEXT_FIELD_NAME), dealerSearchLink: this.getFieldValueForPhraseId(DEALER_SEARCH_PHRASE_ID, LINK_FIELD_NAME), sendMetaFunctionClickAnalyticsEvent: this.sendMetaFunctionClickAnalyticsEvent }),
            h("phn-myporsche", { myporscheText: this.getFieldValueForPhraseId(MYPORSCHE_PHRASE_ID, TEXT_FIELD_NAME), myporscheLink: this.getFieldValueForPhraseId(MYPORSCHE_PHRASE_ID, LINK_FIELD_NAME), loggedInStatusTranslations: this.getLoggedInStatusTranslations(), sendMetaFunctionClickAnalyticsEvent: this.sendMetaFunctionClickAnalyticsEvent, setActiveDrawer: this.setActiveDrawer, closeActiveDrawer: this.closeActiveDrawer })));
    }
    static get is() { return "phn-meta-functionalities"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["meta-functionalities.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["meta-functionalities.css"]
    }; }
    static get properties() { return {
        "closeActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(level: number) => void",
                "resolved": "(level: number) => void",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "setActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(element: HTMLElement, level: number) => void",
                "resolved": "(element: HTMLElement, level: number) => void",
                "references": {
                    "HTMLElement": {
                        "location": "global"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
    static get elementRef() { return "hostElement"; }
}
