// @porschehn/navigation/dist/collection/components/crest/crest.js
import { Component, h } from '@stencil/core';
import { CrestClickEvent } from '../../services/datalayer/crest-click-event';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
export class Crest {
    sendCrestClickAnalyticsEvent() {
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window')
        };
        const crestClickEvent = new CrestClickEvent(analyticsSetup);
        const pagData = (window['pagData'] = window['pagData'] || []);
        pagData.push(crestClickEvent.event());
    }
    render() {
        return (h("a", { class: "crest", onClick: () => this.sendCrestClickAnalyticsEvent(), href: "//www.porsche.com/swiss/de/" },
            h("p-marque", { class: "marque" })));
    }
    static get is() { return "phn-crest"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["crest.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["crest.css"]
    }; }
}
