// @porschehn/navigation/dist/collection/components/shops-and-more-overview/shops-and-more-overview.js
import { Component, Method, Prop, h } from '@stencil/core';
import { hasLinkItemChild } from '../entities/item';
import { EMPTY_ITEM } from '../../services/constants';
import { NavigationLoadEvent } from '../../services/datalayer/navigation-load-event';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
export class ShopsAndMoreOverview {
    constructor() {
        this.item = EMPTY_ITEM;
    }
    async triggerAnalyticsEvent() {
        const pagData = (window['pagData'] = window['pagData'] || []);
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window'),
            environment: analyticsSetupStore.get('environment')
        };
        const analyticsEvent = new NavigationLoadEvent(analyticsSetup, this.item);
        pagData.push(analyticsEvent.event());
    }
    renderLinkGroup(_items, _title) {
        return (h("div", { class: "link-group" },
            h("phn-headline", { text: _title, class: "head-line" }),
            h("div", { class: "list-item-group" }, _items.map((linkItem) => {
                return h("phn-list-item", { item: linkItem });
            }))));
    }
    renderLinkGroups(_items) {
        return (_items &&
            _items.map((childItem) => {
                return this.renderLinkGroup(childItem.Items, childItem.Text);
            }));
    }
    render() {
        return (h("div", { class: "shops-and-more-overview" },
            h("div", { class: "shops-and-more-overview-content" }, hasLinkItemChild(this.item)
                ? this.renderLinkGroup(this.item.Items, this.item.Text)
                : this.renderLinkGroups(this.item.Items))));
    }
    static get is() { return "phn-shops-and-more-overview"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["shops-and-more-overview.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["shops-and-more-overview.css"]
    }; }
    static get properties() { return {
        "item": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ContentItem",
                "resolved": "ContentItem",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_ITEM"
        }
    }; }
    static get methods() { return {
        "triggerAnalyticsEvent": {
            "complexType": {
                "signature": "() => Promise<void>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "",
                "tags": []
            }
        }
    }; }
}
