// @porschehn/navigation/dist/collection/components/menu-drawer-level-1/menu-drawer-level-1.js
import { Component, Element, Event, h, Listen, Method } from '@stencil/core';
import { MenuDrawerAnimation } from '../../services/menu-drawer-animation';
import { navContentStore } from '../../services/nav-content-store';
export class MenuDrawerLevel1 {
    constructor() {
        this.menuDrawerAnimation = new MenuDrawerAnimation();
        this.activeDrawerLevel = 1;
        this.setActiveDrawer = (element, level) => {
            this.activeDrawerElement = element;
            this.activeDrawerLevel = level;
            this.scrollActiveElementToTop();
            this.preventiOSScrollingParentDrawer(level, element, true);
        };
        this.closeActiveDrawer = (level) => {
            const drawerToBeClosed = this.activeDrawerElement;
            this.preventiOSScrollingParentDrawer(level, drawerToBeClosed, false);
            this.activeDrawerLevel = level - 1;
            this.activeDrawerElement = this.el.querySelector(`.drawer-level-${this.activeDrawerLevel}-item`);
            this.scrollActiveElementToTop();
        };
        this.preventiOSScrollingParentDrawer = (targetLevelForAction, targetElementForAction, isLevelIncrease) => {
            /* This prevents iOS allowing the bottom drawer to be scrolled.
             * At the time of creating this solution, we tested overflow: hidden in the fixed scrollable div,
             * as well as pointer-events: none, but this did not work. Somehow, putting
             * position: fixed to the parent drawer container did even if it's not a scrollable container.
             */
            const parentLevel = targetLevelForAction - 1;
            if (isLevelIncrease) {
                const parentScrollableContainer = targetElementForAction.closest(`.drawer-level-${parentLevel}-item`);
                if (parentScrollableContainer) {
                    parentScrollableContainer.querySelector('.drawer-container').style.position = 'fixed';
                }
            }
            else {
                const parentScrollableContainer = targetElementForAction.closest(`.drawer-level-${parentLevel}-item`);
                parentScrollableContainer.querySelector('.drawer-container').style.position = '';
            }
        };
        this.clickBackground = () => {
            const closeButton = this.activeDrawerElement.querySelector('phn-close-button');
            closeButton.click();
        };
        this.clickMenu = (event) => {
            event.stopPropagation();
        };
    }
    async show() {
        this.menuOverlayOpen.emit();
        this.menuDrawerAnimation.slideIn(this.el);
        this.setActiveDrawer(this.el, 1);
    }
    handleClose() {
        this.mobileCarSeriesListElement.hideAll();
        this.hide();
        this.el.querySelectorAll('.drawer-container').forEach((element) => {
            element.style.position = '';
        });
    }
    scrollActiveElementToTop() {
        let scrollableElement;
        if (this.activeDrawerElement.classList.contains(`drawer-level-${this.activeDrawerLevel}-item-scroll`)) {
            scrollableElement = this.activeDrawerElement;
        }
        else {
            scrollableElement = this.activeDrawerElement.querySelector(`.drawer-level-${this.activeDrawerLevel}-item-scroll`);
        }
        if (scrollableElement) {
            scrollableElement.scrollTop = 0;
        }
    }
    hide() {
        this.menuOverlayClose.emit();
        this.menuDrawerAnimation.slideOut(this.el);
    }
    componentDidLoad() {
        this.hide();
    }
    render() {
        const seriesList = navContentStore.get('seriesList');
        return (h("div", { class: "menu-drawer-level-1 drawer-level-1-item", onClick: this.clickBackground },
            h("div", { class: "drawer-level-1 drawer-level-1-item-scroll", onClick: this.clickMenu },
                h("div", { class: "drawer-container" },
                    h("div", { class: "drawer-close" },
                        h("phn-close-button", { class: "close-button" })),
                    h("div", { class: "drawer-content" },
                        h("phn-mobile-car-series-list", { ref: (el) => (this.mobileCarSeriesListElement = el), class: "mobile-car-series-list", carSeries: seriesList, setActiveDrawer: this.setActiveDrawer, closeActiveDrawer: this.closeActiveDrawer }),
                        h("phn-mobile-shops-and-more", { class: "mobile-shops-and-more", setActiveDrawer: this.setActiveDrawer, closeActiveDrawer: this.closeActiveDrawer }),
                        h("phn-meta-functionalities", { class: "meta-functionalities", setActiveDrawer: this.setActiveDrawer, closeActiveDrawer: this.closeActiveDrawer }))))));
    }
    static get is() { return "phn-menu-drawer-level-1"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["menu-drawer-level-1.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["menu-drawer-level-1.css"]
    }; }
    static get events() { return [{
            "method": "menuOverlayOpen",
            "name": "menuOverlayOpen",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "menuOverlayClose",
            "name": "menuOverlayClose",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get methods() { return {
        "show": {
            "complexType": {
                "signature": "() => Promise<void>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "",
                "tags": []
            }
        }
    }; }
    static get elementRef() { return "el"; }
    static get listeners() { return [{
            "name": "closeEvent",
            "method": "handleClose",
            "target": undefined,
            "capture": false,
            "passive": false
        }]; }
}
