// @porsche-design-system/components-js/dist/cjs/p-divider.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const breakpointCustomizable = require('./breakpointCustomizable-2c3f2e06.js');
const index = require('./index-638bfbea.js');

const dividerCss = ":host{display:block}.p-divider{margin:0;padding:0;border:none;text-align:left}.p-divider--color-neutral-contrast-high.p-divider--theme-light{background-color:#323639}.p-divider--color-neutral-contrast-high.p-divider--theme-dark{background-color:#e3e4e5}.p-divider--color-neutral-contrast-medium.p-divider--theme-light{background-color:#626669}.p-divider--color-neutral-contrast-medium.p-divider--theme-dark{background-color:#b0b1b2}.p-divider--color-neutral-contrast-low.p-divider--theme-light{background-color:#c9cacb}.p-divider--color-neutral-contrast-low.p-divider--theme-dark{background-color:#4a4e51}.p-divider--orientation-horizontal{height:1px;width:100%}.p-divider--orientation-vertical{height:100%;width:1px}@media (min-width: 480px){.p-divider--orientation-horizontal-xs{height:1px;width:100%}.p-divider--orientation-vertical-xs{height:100%;width:1px}}@media (min-width: 760px){.p-divider--orientation-horizontal-s{height:1px;width:100%}.p-divider--orientation-vertical-s{height:100%;width:1px}}@media (min-width: 1000px){.p-divider--orientation-horizontal-m{height:1px;width:100%}.p-divider--orientation-vertical-m{height:100%;width:1px}}@media (min-width: 1300px){.p-divider--orientation-horizontal-l{height:1px;width:100%}.p-divider--orientation-vertical-l{height:100%;width:1px}}@media (min-width: 1760px){.p-divider--orientation-horizontal-xl{height:1px;width:100%}.p-divider--orientation-vertical-xl{height:100%;width:1px}}";

class Divider {
    constructor(hostRef) {
        breakpointCustomizable.registerInstance(this, hostRef);
        /** Defines color depending on theme. */
        this.color = 'neutral-contrast-low';
        /** Defines orientation. */
        this.orientation = 'horizontal';
        /** Adapts color depending on theme. */
        this.theme = 'light';
    }
    render() {
        const dividerClasses = index.classnames(breakpointCustomizable.prefix('divider'), breakpointCustomizable.prefix(`divider--color-${this.color}`), breakpointCustomizable.prefix(`divider--theme-${this.theme}`), breakpointCustomizable.mapBreakpointPropToPrefixedClasses('divider--orientation', this.orientation));
        return (breakpointCustomizable.h("hr", { class: dividerClasses }));
    }
}
Divider.style = dividerCss;

exports.p_divider = Divider;
