// @porsche-design-system/components-js/dist/cjs/p-text-list.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-b721d08d.js');
const index$1 = require('./index-66ba56ab.js');

const textListCss = ":host{display:block;counter-reset:section !important}.p-text-list{font-size:1rem;line-height:1.5;font-family:\"Porsche Next\", \"Arial Narrow\", Arial, sans-serif;font-weight:400;display:block;padding:0;margin:0}.p-text-list--theme-light{color:#000}.p-text-list--theme-dark{color:#fff}";

const TextList = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** The type of the text list. */
        this.listType = 'unordered';
        /** The list style type of an ordered list. */
        this.orderType = 'numbered';
        /** Adapts the text color depending on the theme. Has no effect when "inherit" is set as color prop. */
        this.theme = 'light';
    }
    render() {
        const TagType = this.listType === 'unordered' ? 'ul' : 'ol';
        const textListClasses = index$1.cx(index.prefix('text-list'), index.prefix(`text-list--${this.listType}`), index.prefix(`text-list--theme-${this.theme}`), this.isNestedList && index.prefix('text-list--nested'));
        return (index.h(index.Host, { nested: this.isNestedList }, index.h(TagType, { role: 'list', class: textListClasses }, index.h("slot", null))));
    }
    get isNestedList() {
        return !!this.host.closest(index.prefix('text-list-item'));
    }
    get host() { return index.getElement(this); }
};
TextList.style = textListCss;

exports.p_text_list = TextList;
