// @porsche-design-system/components-js/dist/collection/components/navigation/link/link.js
import { Component, Element, h, Prop } from '@stencil/core';
import cx from 'classnames';
import { mapBreakpointPropToPrefixedClasses, prefix } from '../../../utils';
import { improveFocusHandlingForCustomElement } from '../../../utils/focusHandling';
import { insertSlottedStyles } from '../../../utils/slotted-styles';
export class Link {
    constructor() {
        /** The style variant of the link. */
        this.variant = 'secondary';
        /** The icon shown. */
        this.icon = 'arrow-head-right';
        /** A custom URL path to a custom icon. */
        this.iconSource = undefined;
        /** When providing an url then the component will be rendered as `<a>`. */
        this.href = undefined;
        /** Adapts the link color when used on dark background. */
        this.theme = 'light';
        /** Target attribute where the link should be opened. */
        this.target = '_self';
        /** Special download attribute to open native browser download dialog if target url points to a downloadable file. */
        this.download = undefined;
        /** Specifies the relationship of the target object to the link object. */
        this.rel = undefined;
        /** Show or hide label. For better accessibility it is recommended to show the label. */
        this.hideLabel = false;
    }
    componentDidLoad() {
        const tagName = this.element.tagName.toLowerCase();
        const style = `a:focus ${tagName} {
      outline: 2px solid #00d5b9;
      outline-offset: 1px;
    }`;
        insertSlottedStyles(this.element, style);
        improveFocusHandlingForCustomElement(this.element);
    }
    render() {
        const TagType = this.href === undefined ? 'span' : 'a';
        const linkClasses = cx(prefix('link'), prefix(`link--${this.variant}`), mapBreakpointPropToPrefixedClasses('link-', this.hideLabel, ['without-label', 'with-label']), prefix(`link--theme-${this.theme}`));
        const iconClasses = prefix('link__icon');
        const labelClasses = prefix('link__label');
        return (h(TagType, Object.assign({ class: linkClasses }, (TagType === 'a' ? {
            href: this.href,
            target: `${this.target}`,
            download: this.download,
            rel: this.rel
        } : null)),
            h("p-icon", { class: iconClasses, size: 'inherit', name: this.icon, source: this.iconSource, color: 'inherit', "aria-hidden": 'true' }),
            h("p-text", { tag: 'span', color: 'inherit', class: labelClasses },
                h("slot", null))));
    }
    static get is() { return "p-link"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["link.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["link.css"]
    }; }
    static get properties() { return {
        "variant": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "'primary' | 'secondary' | 'tertiary'",
                "resolved": "\"primary\" | \"secondary\" | \"tertiary\"",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "The style variant of the link."
            },
            "attribute": "variant",
            "reflect": false,
            "defaultValue": "'secondary'"
        },
        "icon": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "IconName",
                "resolved": "\"light\" | \"reset\" | \"broadcast\" | \"car\" | \"check\" | \"city\" | \"close\" | \"compare\" | \"country-road\" | \"cubic-capacity\" | \"delete\" | \"document\" | \"download\" | \"edit\" | \"email\" | \"filter\" | \"globe\" | \"highway\" | \"leaf\" | \"locate\" | \"logout\" | \"pause\" | \"phone\" | \"pin\" | \"play\" | \"replay\" | \"rss\" | \"save\" | \"search\" | \"share\" | \"snowflake\" | \"sort\" | \"tablet\" | \"upload\" | \"video\" | \"weight\" | \"wrench\" | \"360\" | \"add\" | \"adjust\" | \"augmented-reality\" | \"calculator\" | \"calender\" | \"camera\" | \"card\" | \"charging-station\" | \"chart\" | \"chat\" | \"clock\" | \"closed-caption\" | \"co2-emission\" | \"configurate\" | \"disable\" | \"duration\" | \"exclamation\" | \"external\" | \"flash\" | \"fuel-station\" | \"gift\" | \"grid\" | \"home\" | \"image\" | \"increase\" | \"information\" | \"leather\" | \"list\" | \"lock\" | \"mobile\" | \"moon\" | \"oil-can\" | \"plug\" | \"printer\" | \"purchase\" | \"question\" | \"racing-flag\" | \"refresh\" | \"route\" | \"screen\" | \"shopping-cart\" | \"stack\" | \"star\" | \"steering-wheel\" | \"stopwatch\" | \"subtract\" | \"sun\" | \"switch\" | \"tachometer\" | \"truck\" | \"user-group\" | \"user\" | \"view\" | \"warning\" | \"wifi\" | \"work\" | \"arrow-double-down\" | \"arrow-double-left\" | \"arrow-double-right\" | \"arrow-down\" | \"arrow-first\" | \"arrow-head-down\" | \"arrow-head-left\" | \"arrow-head-right\" | \"arrow-head-up\" | \"arrow-last\" | \"arrow-up\" | \"battery-empty\" | \"logo-baidu\" | \"logo-delicious\" | \"logo-digg\" | \"logo-facebook\" | \"logo-foursquare\" | \"logo-gmail\" | \"logo-hatena\" | \"logo-instagram\" | \"logo-kaixin\" | \"logo-linkedin\" | \"logo-pinterest\" | \"logo-qq-share\" | \"logo-qq\" | \"logo-skyrock\" | \"logo-sohu\" | \"logo-tecent\" | \"logo-telegram\" | \"logo-tumblr\" | \"logo-twitter\" | \"logo-viber\" | \"logo-vk\" | \"logo-wechat\" | \"logo-weibo\" | \"logo-whatsapp\" | \"logo-xing\" | \"logo-yahoo\" | \"logo-youku\" | \"logo-youtube\" | \"menu-dots-horizontal\" | \"menu-lines\" | \"view-off\" | \"volume-off\" | \"volume-up\" | \"zoom-in\" | \"zoom-out\" | \"lock-open\" | \"arrow-double-up\" | \"arrow-left\" | \"arrow-right\" | \"minus\" | \"plus\" | \"logo-google\"",
                "references": {
                    "IconName": {
                        "location": "import",
                        "path": "../../icon/icon/icon-name"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "The icon shown."
            },
            "attribute": "icon",
            "reflect": false,
            "defaultValue": "'arrow-head-right'"
        },
        "iconSource": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "A custom URL path to a custom icon."
            },
            "attribute": "icon-source",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "href": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "When providing an url then the component will be rendered as `<a>`."
            },
            "attribute": "href",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "theme": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "Theme",
                "resolved": "\"dark\" | \"light\"",
                "references": {
                    "Theme": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Adapts the link color when used on dark background."
            },
            "attribute": "theme",
            "reflect": false,
            "defaultValue": "'light'"
        },
        "target": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "LinkTarget",
                "resolved": "string",
                "references": {
                    "LinkTarget": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Target attribute where the link should be opened."
            },
            "attribute": "target",
            "reflect": false,
            "defaultValue": "'_self'"
        },
        "download": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Special download attribute to open native browser download dialog if target url points to a downloadable file."
            },
            "attribute": "download",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "rel": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Specifies the relationship of the target object to the link object."
            },
            "attribute": "rel",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "hideLabel": {
            "type": "any",
            "mutable": false,
            "complexType": {
                "original": "BreakpointCustomizable<boolean>",
                "resolved": "BreakpointValues<boolean> | boolean | string",
                "references": {
                    "BreakpointCustomizable": {
                        "location": "import",
                        "path": "../../../utils"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Show or hide label. For better accessibility it is recommended to show the label."
            },
            "attribute": "hide-label",
            "reflect": false,
            "defaultValue": "false"
        }
    }; }
    static get elementRef() { return "element"; }
}
