// @porschehn/navigation/dist/collection/components/burger-button/burger-button.js
import { Component, Host, h } from '@stencil/core';
import { navContentStore } from '../../services/nav-content-store';
import { MoreClickEvent } from '../../services/datalayer/more-click-event';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
import { MENU_PHRASE_ID } from '../../services/constants';
export class BurgerButton {
    getText() {
        let text = '';
        const menu = navContentStore.get('menu');
        if (menu && menu.length > 0) {
            const menuItem = menu.find((m) => m.PhraseId === MENU_PHRASE_ID);
            if (menuItem && menuItem.Text) {
                text = menuItem.Text;
            }
        }
        return text;
    }
    sendMoreClickAnalyticsEvent() {
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window'),
            environment: analyticsSetupStore.get('environment')
        };
        const menu = navContentStore.get('menu');
        const moreClickEvent = new MoreClickEvent(analyticsSetup, menu[0]);
        const pagData = (window['pagData'] = window['pagData'] || []);
        pagData.push(moreClickEvent.event());
    }
    render() {
        return (h(Host, { class: "burger-button", onClick: () => this.sendMoreClickAnalyticsEvent() },
            h("p-text", { size: "inherit", class: "burger-button-text" }, this.getText()),
            h("p-icon", { class: "burger-button-icon", size: "small", name: "menu-lines" })));
    }
    static get is() { return "phn-burger-button"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["burger-button.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["burger-button.css"]
    }; }
}
