// @porschehn/navigation/dist/collection/components/burger-button/burger-button.js
import { Component, Host, Prop, h } from '@stencil/core';
import StateTunnel from '../../services/state-tunnel';
import { MoreClickEvent } from '../../services/datalayer/more-click-event';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
import { MENU_PHRASE_ID } from '../../services/constants';
export class BurgerButton {
    constructor() {
        this.menu = [];
    }
    getText() {
        let text = '';
        if (this.menu && this.menu.length > 0) {
            const menuItem = this.menu.find((m) => m.PhraseId === MENU_PHRASE_ID);
            if (menuItem && menuItem.Text) {
                text = menuItem.Text;
            }
        }
        return text;
    }
    sendMoreClickAnalyticsEvent() {
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window')
        };
        const moreClickEvent = new MoreClickEvent(analyticsSetup, this.menu[0]);
        const pagData = (window['pagData'] = window['pagData'] || []);
        pagData.push(moreClickEvent.event());
    }
    render() {
        return (h(Host, { class: "burger-button", onClick: () => this.sendMoreClickAnalyticsEvent() },
            h("p-text", { size: "inherit", class: "burger-button-text" }, this.getText()),
            h("p-icon", { class: "burger-button-icon", size: "small", name: "menu-lines" })));
    }
    static get is() { return "phn-burger-button"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["burger-button.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["burger-button.css"]
    }; }
    static get properties() { return {
        "menu": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ContentItem[]",
                "resolved": "ContentItem[]",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../../services/state-tunnel"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "[]"
        }
    }; }
}
StateTunnel.injectProps(BurgerButton, ['menu']);
