// @porschehn/navigation/dist/collection/components/mobile-car-series-list/mobile-car-series-list.js
import { Component, h, Prop } from '@stencil/core';
import { DEFAULT_SERIES_LIST, FEATURE_NEW_LEVEL2 } from '../../utility/constants';
import { navStateStore } from '../../state/nav-state-store';
import { analyticsSetupStore } from '../../state/analytics-setup-store';
import { ModelRangeClickEvent } from '../../services/datalayer/model-range-click-event';
import { isFeatureEnabled } from '../../env-config';
export class MobileCarSeriesList {
    constructor() {
        this.carSeries = DEFAULT_SERIES_LIST;
        this.sendNavigationClickAnalyticsEvent = (series) => {
            const analyticsSetup = {
                applicationId: analyticsSetupStore.get('applicationId'),
                locale: analyticsSetupStore.get('locale'),
                window: analyticsSetupStore.get('window'),
                environment: analyticsSetupStore.get('environment')
            };
            const modelSeriesClickEvent = new ModelRangeClickEvent(analyticsSetup, series.Name, series);
            const pagData = (window['pagData'] = window['pagData'] || []);
            pagData.push(modelSeriesClickEvent.event());
        };
    }
    showDrawer(_, series, index) {
        if (isFeatureEnabled(FEATURE_NEW_LEVEL2)) {
            const phnCarSeriesModelOverviewV2 = {
                content: (h("phn-car-series-model-overview-v2", { series: series, class: 'drawer-car-series-model-overview-' + index, slot: "content" }))
            };
            navStateStore.set('drawerContentStack', [
                ...navStateStore.get('drawerContentStack'),
                phnCarSeriesModelOverviewV2
            ]);
            this.sendNavigationClickAnalyticsEvent(series);
        }
        else {
            const phnDrawerCarSeriesModelOverview = {
                content: (h("phn-drawer-car-series-model-overview", { series: series, class: 'drawer-car-series-model-overview drawer-car-series-model-overview-' + index, slot: "content" }))
            };
            navStateStore.set('drawerContentStack', [
                ...navStateStore.get('drawerContentStack'),
                phnDrawerCarSeriesModelOverview
            ]);
            this.sendNavigationClickAnalyticsEvent(series);
        }
    }
    render() {
        return (h("div", { class: "mobile-car-series-list" }, this.carSeries.map((series, index) => (h("div", { class: "mobile-car-series", key: index },
            h("button", { class: "car-series-button", onClick: (event) => this.showDrawer(event, series, index) },
                h("phn-car-series", { class: 'car-series car-series-' + index, text: series.Name, image: series.Image })))))));
    }
    static get is() { return "phn-mobile-car-series-list"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["mobile-car-series-list.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["mobile-car-series-list.css"]
    }; }
    static get properties() { return {
        "carSeries": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Series[]",
                "resolved": "Series[]",
                "references": {
                    "Series": {
                        "location": "import",
                        "path": "../../entities/series"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "DEFAULT_SERIES_LIST"
        }
    }; }
}
