// @porschehn/navigation/dist/collection/components/myporsche/myporsche.js
import { Component, Host, Prop, h } from '@stencil/core';
import { FEATURE_DRAWER_RESTRUCTURE, LOADING_STATUS, LOGGED_OUT_USER_STATUS, MYPORSCHE_PHRASE_ID } from '../../services/constants';
import { navContentStore } from '../../services/nav-content-store';
import { isFeatureEnabled } from '../../env-config';
export class Myporsche {
    constructor() {
        this.myporscheLink = '';
        this.myporscheText = '';
        this.loggedInStatusTranslations = {};
    }
    getSubTextForUser() {
        const loggedInState = navContentStore.get('loggedInState');
        const phraseId = `${MYPORSCHE_PHRASE_ID}.${loggedInState}`;
        return this.loggedInStatusTranslations[phraseId] || loggedInState;
    }
    renderLoggedOutComponent() {
        return (h("p-link-pure", { class: "myporsche-logged-out", href: this.myporscheLink, onClick: () => {
                const linkItem = {
                    PhraseId: MYPORSCHE_PHRASE_ID,
                    Text: this.myporscheText,
                    Link: this.myporscheLink
                };
                this.sendMetaFunctionClickAnalyticsEvent(linkItem);
            }, size: "inherit", icon: "user" },
            this.myporscheText,
            h("p-text", { class: "login-status", color: "inherit", size: "inherit", ellipsis: "true", align: "left" }, this.getSubTextForUser())));
    }
    renderLoadingComponent() {
        return (h("p-button-pure", { loading: "true", size: "inherit" }, this.myporscheText));
    }
    renderLoggedInComponent() {
        const unreadMessagesCount = navContentStore.get('unreadMessagesCount');
        const drawerRestructure = isFeatureEnabled(FEATURE_DRAWER_RESTRUCTURE, navContentStore.get('env'));
        const myPorscheDrawer = drawerRestructure ? (h("phn-myporsche-drawer-2", { class: "myporsche-logged-in", myPorscheText: this.myporscheText, loginStatus: this.getSubTextForUser(), unreadMessagesCount: unreadMessagesCount })) : (h("phn-myporsche-drawer", { class: "myporsche-logged-in", myPorscheText: this.myporscheText, loginStatus: this.getSubTextForUser(), setActiveDrawer: this.setActiveDrawer, closeActiveDrawer: this.closeActiveDrawer, unreadMessagesCount: unreadMessagesCount }));
        return [
            myPorscheDrawer,
            h("phn-myporsche-flyout", { class: "myporsche-flyout", myPorscheText: this.myporscheText, loginStatus: this.getSubTextForUser(), unreadMessagesCount: unreadMessagesCount })
        ];
    }
    render() {
        const loggedInState = navContentStore.get('loggedInState');
        return (h(Host, null, loggedInState === LOGGED_OUT_USER_STATUS
            ? this.renderLoggedOutComponent()
            : loggedInState === LOADING_STATUS
                ? this.renderLoadingComponent()
                : this.renderLoggedInComponent()));
    }
    static get is() { return "phn-myporsche"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["myporsche.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["myporsche.css"]
    }; }
    static get properties() { return {
        "myporscheLink": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "myporsche-link",
            "reflect": false,
            "defaultValue": "''"
        },
        "myporscheText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "myporsche-text",
            "reflect": false,
            "defaultValue": "''"
        },
        "sendMetaFunctionClickAnalyticsEvent": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(linkItem: ContentItem) => void",
                "resolved": "(linkItem: ContentItem) => void",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "loggedInStatusTranslations": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "{}",
                "resolved": "{}",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "{}"
        },
        "closeActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(level: number) => void",
                "resolved": "(level: number) => void",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "setActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(element: HTMLElement, level: number) => void",
                "resolved": "(element: HTMLElement, level: number) => void",
                "references": {
                    "HTMLElement": {
                        "location": "global"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
}
