// @porschehn/navigation/dist/cjs/app-globals-4d9a31f5.js
'use strict';

const _commonjsHelpers = require('./_commonjsHelpers-ce3c415c.js');

/*
 Stencil Client Patch v1.16.5 | MIT Licensed | https://stenciljs.com
 */
const noop = () => {
    /* noop*/
};
const IS_DENO_ENV = typeof Deno !== 'undefined';
const IS_NODE_ENV = !IS_DENO_ENV &&
    typeof global !== 'undefined' &&
    typeof require === 'function' &&
    !!global.process &&
    typeof __filename === 'string' &&
    (!global.origin || typeof global.origin !== 'string');
const IS_DENO_WINDOWS_ENV = IS_DENO_ENV && Deno.build.os === 'windows';
const getCurrentDirectory = IS_NODE_ENV ? process.cwd : IS_DENO_ENV ? Deno.cwd : () => '/';
const exit = IS_NODE_ENV ? process.exit : IS_DENO_ENV ? Deno.exit : noop;
const getDynamicImportFunction = (namespace) => `__sc_import_${namespace.replace(/\s|-/g, '_')}`;
const patchEsm = () => {
    // NOTE!! This fn cannot use async/await!
    // @ts-ignore
    if ( !(_commonjsHelpers.CSS && _commonjsHelpers.CSS.supports && _commonjsHelpers.CSS.supports('color', 'var(--c)'))) {
        // @ts-ignore
        return Promise.resolve().then(function () { return require(/* webpackChunkName: "polyfills-css-shim" */ './css-shim-74b53181.js'); }).then(() => {
            if ((_commonjsHelpers.plt.$cssShim$ = _commonjsHelpers.win.__cssshim)) {
                return _commonjsHelpers.plt.$cssShim$.i();
            }
            else {
                // for better minification
                return 0;
            }
        });
    }
    return _commonjsHelpers.promiseResolve();
};
const patchBrowser = () => {
    {
        // shim css vars
        _commonjsHelpers.plt.$cssShim$ = _commonjsHelpers.win.__cssshim;
    }
    // @ts-ignore
    const scriptElm =  Array.from(_commonjsHelpers.doc.querySelectorAll('script')).find(s => new RegExp(`\/${_commonjsHelpers.NAMESPACE}(\\.esm)?\\.js($|\\?|#)`).test(s.src) || s.getAttribute('data-stencil-namespace') === _commonjsHelpers.NAMESPACE)
        ;
    const opts =  scriptElm['data-opts'] || {} ;
    if ( 'onbeforeload' in scriptElm && !history.scrollRestoration /* IS_ESM_BUILD */) {
        // Safari < v11 support: This IF is true if it's Safari below v11.
        // This fn cannot use async/await since Safari didn't support it until v11,
        // however, Safari 10 did support modules. Safari 10 also didn't support "nomodule",
        // so both the ESM file and nomodule file would get downloaded. Only Safari
        // has 'onbeforeload' in the script, and "history.scrollRestoration" was added
        // to Safari in v11. Return a noop then() so the async/await ESM code doesn't continue.
        // IS_ESM_BUILD is replaced at build time so this check doesn't happen in systemjs builds.
        return {
            then() {
                /* promise noop */
            },
        };
    }
    {
        opts.resourcesUrl = new URL('.', new URL(scriptElm.getAttribute('data-resources-url') || scriptElm.src, _commonjsHelpers.win.location.href)).href;
        {
            patchDynamicImport(opts.resourcesUrl, scriptElm);
        }
        if ( !_commonjsHelpers.win.customElements) {
            // module support, but no custom elements support (Old Edge)
            // @ts-ignore
            return Promise.resolve().then(function () { return require(/* webpackChunkName: "polyfills-dom" */ './dom-da5f340b.js'); }).then(() => opts);
        }
    }
    return _commonjsHelpers.promiseResolve(opts);
};
const patchDynamicImport = (base, orgScriptElm) => {
    const importFunctionName = getDynamicImportFunction(_commonjsHelpers.NAMESPACE);
    try {
        // test if this browser supports dynamic imports
        // There is a caching issue in V8, that breaks using import() in Function
        // By generating a random string, we can workaround it
        // Check https://bugs.chromium.org/p/chromium/issues/detail?id=990810 for more info
        _commonjsHelpers.win[importFunctionName] = new Function('w', `return import(w);//${Math.random()}`);
    }
    catch (e) {
        // this shim is specifically for browsers that do support "esm" imports
        // however, they do NOT support "dynamic" imports
        // basically this code is for old Edge, v18 and below
        const moduleMap = new Map();
        _commonjsHelpers.win[importFunctionName] = (src) => {
            const url = new URL(src, base).href;
            let mod = moduleMap.get(url);
            if (!mod) {
                const script = _commonjsHelpers.doc.createElement('script');
                script.type = 'module';
                script.crossOrigin = orgScriptElm.crossOrigin;
                script.src = URL.createObjectURL(new Blob([`import * as m from '${url}'; window.${importFunctionName}.m = m;`], { type: 'application/javascript' }));
                mod = new Promise(resolve => {
                    script.onload = () => {
                        resolve(_commonjsHelpers.win[importFunctionName].m);
                        script.remove();
                    };
                });
                moduleMap.set(url, mod);
                _commonjsHelpers.doc.head.appendChild(script);
            }
            return mod;
        };
    }
};

var componentsJs = _commonjsHelpers.createCommonjsModule(function (module, exports) {
!function(e,t){module.exports=t();}("undefined"!=typeof self?self:_commonjsHelpers.commonjsGlobal,(function(){return function(){var e={d:function(t,n){for(var o in n)e.o(n,o)&&!e.o(t,o)&&Object.defineProperty(t,o,{enumerable:!0,get:n[o]});},o:function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r:function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0});}},t={};function n(e){if("noModule"in HTMLScriptElement.prototype){const t=(document.createElement("script"));t.src=e,t.setAttribute("crossorigin",""),document.body.appendChild(t);}}e.r(t),e.d(t,{componentsReady:function(){return s},load:function(){return r}});const o="porscheDesignSystem";const r=(e={prefix:""})=>{!function({script:e,version:t,prefix:r}){var s;const c=null!==(s=function(e){const t=(document[o]||(document[o]={}),document[o]),{[e]:n=null}=t;if(null===n){const n={isLoaded:!1,prefixes:[],registerCustomElements:null};t[e]=n;}return t[e]}(t))&&void 0!==s?s:{},{isLoaded:i,prefixes:d,registerCustomElements:u}=c;i||(n(e),c.isLoaded=!0),d.includes(r)||(d.push(r),u&&u(r));}(Object.assign(Object.assign({},{version:"2.0.2",script:(typeof window!=='undefined'&&window.PORSCHE_DESIGN_SYSTEM_CDN==='cn'?'https://cdn.ui.porsche.cn':'https://cdn.ui.porsche.com')+"/porsche-design-system/components/porsche-design-system.v2.0.2.af5ac9ef7e24f6353b02.js"}),e));},s=(e=document.body)=>{let t;const n=new Promise((e=>t=e));if(c())i(e,t);else {const n="readystatechange",o=()=>{c()&&(document.removeEventListener(n,o),i(e,t));};document.addEventListener(n,o);}return n},c=()=>"complete"===document.readyState,i=(e,t)=>{const n=d(e);Promise.all(n).then((e=>t(e.length))).catch(console.error);},d=e=>{let t=[];return 1===(null==e?void 0:e.nodeType)&&Array.from(e.children).forEach((e=>{f(e)&&t.push(e.componentOnReady()),t=t.concat(d(e));})),t},u=/^(.*-)?P-(.*)$/,f=e=>u.exec(e.tagName)&&"function"==typeof e.componentOnReady;return t}()}));
});

function appGlobalScript () {
    componentsJs.load({ prefix: 'phn' });
}

const globalScripts = appGlobalScript;

exports.globalScripts = globalScripts;
exports.patchBrowser = patchBrowser;
exports.patchEsm = patchEsm;
