// @porsche-design-system/components-js/dist/collection/components/navigation/link-pure/link-pure.js
import { Component, Element, h, Prop } from '@stencil/core';
import cx from 'classnames';
import { calcLineHeightForElement, mapBreakpointPropToPrefixedClasses, prefix, transitionListener, insertSlottedStyles } from '../../../utils';
import { improveFocusHandlingForCustomElement } from '../../../utils/focusHandling';
export class LinkPure {
    constructor() {
        /** Size of the link. */
        this.size = 'small';
        /** The weight of the text (only has effect with visible label). */
        this.weight = 'regular';
        /** The icon shown. */
        this.icon = 'arrow-head-right';
        /** A custom URL path to a custom icon. */
        this.iconSource = undefined;
        /** When providing an url then the component will be rendered as `<a>`. */
        this.href = undefined;
        /** Display link in active state. */
        this.active = false;
        /** Show or hide label. For better accessibility it is recommended to show the label. */
        this.hideLabel = false;
        /** Adapts the button color depending on the theme. */
        this.theme = 'light';
        /** Target attribute where the link should be opened. */
        this.target = '_self';
        /** Special download attribute to open native browser download dialog if target url points to a downloadable file. */
        this.download = undefined;
        /** Specifies the relationship of the target object to the link object. */
        this.rel = undefined;
    }
    componentDidLoad() {
        const tagName = this.element.tagName.toLowerCase();
        const style = `a:focus ${tagName} {
      outline: 2px solid #00d5b9;
      outline-offset: 1px;
    }`;
        insertSlottedStyles(this.element, style);
        improveFocusHandlingForCustomElement(this.element);
        transitionListener(this.linkTag, 'font-size', () => {
            const size = calcLineHeightForElement(this.linkTag);
            this.iconTag.style.width = `${size}em`;
            this.iconTag.style.height = `${size}em`;
        });
    }
    render() {
        const TagType = this.href === undefined ? 'span' : 'a';
        const linkPureClasses = cx(prefix('link-pure'), mapBreakpointPropToPrefixedClasses('link-pure--size', this.size), prefix(`link-pure--theme-${this.theme}`), this.active && prefix('link-pure--active'));
        const iconClasses = cx(prefix('link-pure__icon'));
        const labelClasses = cx(prefix('link-pure__label'), mapBreakpointPropToPrefixedClasses('link-pure__label-', this.hideLabel, ['hidden', 'visible']));
        return (h(TagType, Object.assign({ class: linkPureClasses }, (TagType === 'a' ? {
            href: this.href,
            target: this.target,
            download: this.download,
            rel: this.rel
        } : null), { ref: el => this.linkTag = el }),
            h("p-icon", { class: iconClasses, color: 'inherit', size: 'inherit', name: this.icon, source: this.iconSource, ref: el => this.iconTag = el, "aria-hidden": 'true' }),
            h("p-text", { class: labelClasses, tag: 'span', color: 'inherit', size: 'inherit', weight: this.weight },
                h("slot", null))));
    }
    static get is() { return "p-link-pure"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["link-pure.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["link-pure.css"]
    }; }
    static get properties() { return {
        "size": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "BreakpointCustomizable<TextSize>",
                "resolved": "BreakpointValues<TextSize> | string",
                "references": {
                    "BreakpointCustomizable": {
                        "location": "import",
                        "path": "../../../utils"
                    },
                    "TextSize": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Size of the link."
            },
            "attribute": "size",
            "reflect": false,
            "defaultValue": "'small'"
        },
        "weight": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "TextWeight",
                "resolved": "\"bold\" | \"regular\" | \"thin\"",
                "references": {
                    "TextWeight": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "The weight of the text (only has effect with visible label)."
            },
            "attribute": "weight",
            "reflect": false,
            "defaultValue": "'regular'"
        },
        "icon": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "IconName",
                "resolved": "\"light\" | \"reset\" | \"360\" | \"add\" | \"adjust\" | \"arrow-double-down\" | \"arrow-double-left\" | \"arrow-double-right\" | \"arrow-double-up\" | \"arrow-down\" | \"arrow-first\" | \"arrow-head-down\" | \"arrow-head-left\" | \"arrow-head-right\" | \"arrow-head-up\" | \"arrow-last\" | \"arrow-left\" | \"arrow-right\" | \"arrow-up\" | \"augmented-reality\" | \"battery-empty\" | \"broadcast\" | \"calculator\" | \"calendar\" | \"camera\" | \"car\" | \"card\" | \"charging-station\" | \"chart\" | \"chat\" | \"check\" | \"city\" | \"clock\" | \"close\" | \"closed-caption\" | \"co2-emission\" | \"compare\" | \"configurate\" | \"country-road\" | \"cubic-capacity\" | \"delete\" | \"disable\" | \"document\" | \"download\" | \"duration\" | \"edit\" | \"email\" | \"exclamation\" | \"external\" | \"filter\" | \"flash\" | \"fuel-station\" | \"gift\" | \"globe\" | \"grid\" | \"highway\" | \"home\" | \"image\" | \"increase\" | \"information\" | \"leaf\" | \"leather\" | \"list\" | \"locate\" | \"lock\" | \"lock-open\" | \"logo-baidu\" | \"logo-delicious\" | \"logo-digg\" | \"logo-facebook\" | \"logo-foursquare\" | \"logo-gmail\" | \"logo-google\" | \"logo-hatena\" | \"logo-instagram\" | \"logo-kaixin\" | \"logo-linkedin\" | \"logo-pinterest\" | \"logo-qq\" | \"logo-qq-share\" | \"logo-skyrock\" | \"logo-sohu\" | \"logo-tecent\" | \"logo-telegram\" | \"logo-tumblr\" | \"logo-twitter\" | \"logo-viber\" | \"logo-vk\" | \"logo-wechat\" | \"logo-weibo\" | \"logo-whatsapp\" | \"logo-xing\" | \"logo-yahoo\" | \"logo-youku\" | \"logo-youtube\" | \"logout\" | \"menu-dots-horizontal\" | \"menu-lines\" | \"minus\" | \"mobile\" | \"moon\" | \"oil-can\" | \"pause\" | \"phone\" | \"pin\" | \"play\" | \"plug\" | \"plus\" | \"printer\" | \"purchase\" | \"question\" | \"racing-flag\" | \"refresh\" | \"replay\" | \"route\" | \"rss\" | \"save\" | \"screen\" | \"search\" | \"share\" | \"shopping-cart\" | \"snowflake\" | \"sort\" | \"stack\" | \"star\" | \"steering-wheel\" | \"stopwatch\" | \"subtract\" | \"sun\" | \"switch\" | \"tablet\" | \"tachometer\" | \"truck\" | \"upload\" | \"user\" | \"user-group\" | \"video\" | \"view\" | \"view-off\" | \"volume-off\" | \"volume-up\" | \"warning\" | \"weight\" | \"wifi\" | \"work\" | \"wrench\" | \"zoom-in\" | \"zoom-out\"",
                "references": {
                    "IconName": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "The icon shown."
            },
            "attribute": "icon",
            "reflect": false,
            "defaultValue": "'arrow-head-right'"
        },
        "iconSource": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "A custom URL path to a custom icon."
            },
            "attribute": "icon-source",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "href": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "When providing an url then the component will be rendered as `<a>`."
            },
            "attribute": "href",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "active": {
            "type": "boolean",
            "mutable": false,
            "complexType": {
                "original": "boolean",
                "resolved": "boolean",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Display link in active state."
            },
            "attribute": "active",
            "reflect": false,
            "defaultValue": "false"
        },
        "hideLabel": {
            "type": "any",
            "mutable": false,
            "complexType": {
                "original": "BreakpointCustomizable<boolean>",
                "resolved": "BreakpointValues<boolean> | boolean | string",
                "references": {
                    "BreakpointCustomizable": {
                        "location": "import",
                        "path": "../../../utils"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Show or hide label. For better accessibility it is recommended to show the label."
            },
            "attribute": "hide-label",
            "reflect": false,
            "defaultValue": "false"
        },
        "theme": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "Theme",
                "resolved": "\"dark\" | \"light\"",
                "references": {
                    "Theme": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Adapts the button color depending on the theme."
            },
            "attribute": "theme",
            "reflect": false,
            "defaultValue": "'light'"
        },
        "target": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "LinkTarget",
                "resolved": "string",
                "references": {
                    "LinkTarget": {
                        "location": "import",
                        "path": "../../../types"
                    }
                }
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Target attribute where the link should be opened."
            },
            "attribute": "target",
            "reflect": false,
            "defaultValue": "'_self'"
        },
        "download": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Special download attribute to open native browser download dialog if target url points to a downloadable file."
            },
            "attribute": "download",
            "reflect": false,
            "defaultValue": "undefined"
        },
        "rel": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": true,
            "docs": {
                "tags": [],
                "text": "Specifies the relationship of the target object to the link object."
            },
            "attribute": "rel",
            "reflect": false,
            "defaultValue": "undefined"
        }
    }; }
    static get elementRef() { return "element"; }
}
