// @porschehn/navigation/dist/collection/components/myporsche-drawer/myporsche-drawer.js
import { Component, Host, Prop, h } from '@stencil/core';
import { DEFAULT_ENVIRONMENT, FEATURE_KEYBOARD } from '../../services/constants';
import { isFeatureEnabled } from '../../env-config';
export class MyPorscheDrawer {
    constructor() {
        this.env = DEFAULT_ENVIRONMENT;
        this.unreadMessagesCount = 0;
        this.handlePressEnterKey = (event, drawerRef) => {
            if (isFeatureEnabled(FEATURE_KEYBOARD, this.env)) {
                if (event.key === 'Enter') {
                    this.openDrawer(drawerRef);
                }
            }
        };
    }
    openDrawer(myPorscheItems) {
        myPorscheItems.style.display = 'block';
        this.setActiveDrawer(myPorscheItems, 2);
    }
    render() {
        let drawerRef;
        return (h(Host, null,
            h("phn-button-extended", { class: "drawer-trigger", icon: "arrow-head-right", onClick: () => this.openDrawer(drawerRef), onKeyDown: (event) => this.handlePressEnterKey(event, drawerRef), cancelTabIndex: false },
                h("span", { class: this.unreadMessagesCount > 0
                        ? 'myporsche-unread-messages-indicator'
                        : 'myporsche-unread-messages-indicator hidden' }),
                h("p-icon", { name: "user", size: "small" }),
                h("span", { class: "drawer-trigger-label" },
                    this.myPorscheText,
                    h("p-text", { class: "login-status", size: "inherit", ellipsis: "true", align: "left" }, this.loginStatus))),
            h("phn-myporsche-items-drawer", { class: "drawer--closed", ref: (el) => (drawerRef = el), closeActiveDrawer: this.closeActiveDrawer })));
    }
    static get is() { return "phn-myporsche-drawer"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["myporsche-drawer.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["myporsche-drawer.css"]
    }; }
    static get properties() { return {
        "loginStatus": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "login-status",
            "reflect": false
        },
        "myPorscheText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "my-porsche-text",
            "reflect": false
        },
        "closeActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(level: number) => void",
                "resolved": "(level: number) => void",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "setActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(element: HTMLElement, level: number) => void",
                "resolved": "(element: HTMLElement, level: number) => void",
                "references": {
                    "HTMLElement": {
                        "location": "global"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "env": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "env",
            "reflect": false,
            "defaultValue": "DEFAULT_ENVIRONMENT"
        },
        "unreadMessagesCount": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "unread-messages-count",
            "reflect": false,
            "defaultValue": "0"
        }
    }; }
}
