// @porsche-design-system/components-js/dist/esm/p-text-list.entry.js
import { r as registerInstance, d as prefix, h, H as Host, g as getElement } from './index-e8247a5d.js';
import { c as cx } from './index-e6c0a386.js';

const textListCss = ":host{display:block;counter-reset:section !important}.p-text-list{font-size:1rem;line-height:1.5;font-family:\"Porsche Next\", \"Arial Narrow\", Arial, sans-serif;font-weight:400;display:block;padding:0;margin:0}.p-text-list--theme-light{color:#000}.p-text-list--theme-dark{color:#fff}";

const TextList = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        /** The type of the text list. */
        this.listType = 'unordered';
        /** The list style type of an ordered list. */
        this.orderType = 'numbered';
        /** Adapts the text color depending on the theme. Has no effect when "inherit" is set as color prop. */
        this.theme = 'light';
    }
    render() {
        const TagType = this.listType === 'unordered' ? 'ul' : 'ol';
        const textListClasses = cx(prefix('text-list'), prefix(`text-list--${this.listType}`), prefix(`text-list--theme-${this.theme}`), this.isNestedList && prefix('text-list--nested'));
        return (h(Host, { nested: this.isNestedList }, h(TagType, { role: 'list', class: textListClasses }, h("slot", null))));
    }
    get isNestedList() {
        return !!this.host.closest(prefix('text-list-item'));
    }
    get host() { return getElement(this); }
};
TextList.style = textListCss;

export { TextList as p_text_list };
