// @porschehn/navigation/dist/collection/services/body-scroll-lock.js
import * as utils from '../utility/utils';
export class BodyScrollLock {
    constructor() {
        this.activate = () => {
            this.recordCurrentBodyStyles();
            this.initBodyPaddingRight();
            document.body.style.overflow = 'hidden';
            document.body.style.position = 'fixed';
            document.body.style.width = '100%';
            document.body.style.paddingRight = `${this.initialBodyPaddingRightValue + this.scrollbarWidth}${this.initialBodyPaddingRightUnit}`;
        };
        this.deactivate = () => {
            document.body.style.overflow = this.initialBodyOverflow;
            document.body.style.position = this.initialBodyPosition;
            document.body.style.width = this.initialBodyWidth;
            document.body.style.paddingRight = `${this.initialBodyPaddingRightValue}${this.initialBodyPaddingRightUnit}`;
        };
        this.recordCurrentBodyStyles = () => {
            this.initialBodyOverflow = document.body.style.overflow;
            this.initialBodyPosition = document.body.style.position;
            this.initialBodyWidth = document.body.style.width;
        };
        this.scrollbarWidth = utils.setScrollbarWidth();
    }
    initBodyPaddingRight() {
        const body = document.body;
        const paddingRight = window.getComputedStyle(body, null).getPropertyValue('padding-right');
        this.initialBodyPaddingRightValue = parseFloat(paddingRight);
        this.initialBodyPaddingRightUnit = paddingRight.split(this.initialBodyPaddingRightValue.toString())[1];
    }
}
