// @porschehn/navigation/dist/collection/services/constants.js
export const DEFAULT_LOCALE = 'de-CH';
export const DEFAULT_ITEMS_LIST = [];
export const DEFAULT_SERIES_LIST = [];
export const DEFAULT_SHOPS = [];
export const DEFAULT_MORE = [];
export const DEFAULT_METAFUNCTIONALITIES = [];
export const DEFAULT_MENU = [];
export const DEFAULT_ENVIRONMENT = 'ENVIRONMENT';
export const ANALYTICS_APP_ID = 'nav';
export const EMPTY_SERIES = { PhraseId: '', Name: '', Image: '', Models: [], Further: [], Meta: [] };
export const EMPTY_ITEM = { Id: '', PhraseId: '', Text: '', IsLinkItem: false };
export const REDACTED = 'REDACTED';
export const LOGGED_IN_USER_NO_NAME_STATUS = 'loggedIn';
export const LOGGED_OUT_USER_STATUS = 'loginOrRegister';
export const LOADING_STATUS = 'loading';
export const VERSION = 'VERSION_STRING';
export const LINK_FIELD_NAME = 'Link';
export const TEXT_FIELD_NAME = 'Text';
export const DEALER_SEARCH_PHRASE_ID = 'metafunctionalities.finddealer';
export const MYPORSCHE_PHRASE_ID = 'metafunctionalities.myporsche';
export const ACCESS_TOKEN_FETCHED_EVENT = 'accessTokenFetched';
export const ACCESS_TOKEN_FETCH_ERROR_EVENT = 'accessTokenFetchError';
export const FEATURE_OVERRIDE_ENABLED = 'featureOverrideEnabled';
// Defining the feature names
export const FEATURE_FEATURE1 = 'homenav_feature1';
export const FEATURE_KEYBOARD = 'homenav_732_keyboard';
