// @porschehn/navigation/dist/collection/components/back-button/back-button.js
import { Component, Host, Prop, h } from '@stencil/core';
import { analyticsSetupStore } from '../../services/datalayer/analytics-setup-store';
import { MenuBackClickEvent } from '../../services/datalayer/menu-back-click';
export class BackButton {
    constructor() {
        this.text = '';
        this.buttonId = '';
    }
    sendMenuBackClickEventAnalyticsEvent() {
        const contentItem = { Text: this.text, PhraseId: this.buttonId, Link: '' };
        const analyticsSetup = {
            applicationId: analyticsSetupStore.get('applicationId'),
            locale: analyticsSetupStore.get('locale'),
            window: analyticsSetupStore.get('window'),
            environment: analyticsSetupStore.get('environment')
        };
        const menuBackClickEvent = new MenuBackClickEvent(analyticsSetup, contentItem);
        const pagData = (window['pagData'] = window['pagData'] || []);
        pagData.push(menuBackClickEvent.event());
    }
    render() {
        return (h(Host, null,
            h("p-button-pure", { size: "inherit", icon: "arrow-head-left", class: "back-button-text-link", onClick: () => this.sendMenuBackClickEventAnalyticsEvent() }, this.text)));
    }
    static get is() { return "phn-back-button"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["back-button.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["back-button.css"]
    }; }
    static get properties() { return {
        "text": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "text",
            "reflect": false,
            "defaultValue": "''"
        },
        "buttonId": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "button-id",
            "reflect": false,
            "defaultValue": "''"
        }
    }; }
}
