// @porschehn/navigation/dist/collection/components/myporsche-items/myporsche-items.js
import { Component, Host, Prop, h } from '@stencil/core';
import { LinkItem } from '../entities/item';
import StateTunnel from '../../services/state-tunnel';
import { DEFAULT_ENVIRONMENT, MYPORSCHE_PHRASE_ID, FEATURE_KEYBOARD } from '../../services/constants';
import { isFeatureEnabled } from '../../env-config';
export class MyPorscheItems {
    constructor() {
        this.env = DEFAULT_ENVIRONMENT;
        this.handlePressShiftTabKey = (event) => {
            if (isFeatureEnabled(FEATURE_KEYBOARD, this.env)) {
                if (event.shiftKey && event.key === 'Tab') {
                    this.myPorscheCloseHandler();
                }
            }
        };
        this.handlePressTabKey = (event) => {
            if (isFeatureEnabled(FEATURE_KEYBOARD, this.env)) {
                if (!event.shiftKey && event.key === 'Tab') {
                    this.myPorscheCloseHandler();
                }
            }
        };
    }
    getMyPorscheItemFromTemplateAsLinkItem(id) {
        const phraseId = `${MYPORSCHE_PHRASE_ID}.${id}`;
        const myPorscheItems = this.metafunctionalities.find((m) => m.PhraseId === MYPORSCHE_PHRASE_ID).Items;
        const myPorscheItem = myPorscheItems.find((item) => item.Id === phraseId);
        return new LinkItem(myPorscheItem.Id, myPorscheItem.Text, myPorscheItem.Link);
    }
    render() {
        return (h(Host, null,
            h("phn-list-item", { class: "myporsche-list-item", item: this.getMyPorscheItemFromTemplateAsLinkItem('carsandservices'), onKeyDown: this.handlePressShiftTabKey }),
            h("phn-list-item", { class: "myporsche-list-item", item: this.getMyPorscheItemFromTemplateAsLinkItem('messages') }),
            h("phn-list-item", { class: "myporsche-list-item", item: this.getMyPorscheItemFromTemplateAsLinkItem('accountadministration') }),
            h("hr", null),
            h("phn-list-item", { class: "myporsche-list-item logout", item: this.getMyPorscheItemFromTemplateAsLinkItem('logout'), onKeyDown: this.handlePressTabKey })));
    }
    static get is() { return "phn-myporsche-items"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["myporsche-items.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["myporsche-items.css"]
    }; }
    static get properties() { return {
        "metafunctionalities": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ContentItemWithItems[]",
                "resolved": "ContentItemWithItems[]",
                "references": {
                    "ContentItemWithItems": {
                        "location": "import",
                        "path": "../../services/state-tunnel"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "myPorscheCloseHandler": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "() => void",
                "resolved": "() => void",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "env": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "env",
            "reflect": false,
            "defaultValue": "DEFAULT_ENVIRONMENT"
        }
    }; }
}
StateTunnel.injectProps(MyPorscheItems, ['metafunctionalities']);
