// @porschehn/navigation/dist/collection/components/myporsche-flyout/myporsche-flyout.js
import { Component, Host, Listen, Prop, State, h } from '@stencil/core';
import { DEFAULT_ENVIRONMENT } from '../../utility/constants';
export class MyPorscheFlyout {
    constructor() {
        this.env = DEFAULT_ENVIRONMENT;
        this.unreadMessagesCount = 0;
        this.isActive = false;
        this.screenReaderId = 'myporsche-flyout';
        this.keyDownHandler = (event) => {
            if (event.shiftKey && event.key === 'Tab') {
                this.isActive = false;
            }
        };
        // also activates screenreader space and enter
        this.handleClick = (event) => {
            event.stopPropagation(); // prevents button from losing focus
            this.isActive = true;
        };
    }
    myPorscheCloseHandler() {
        this.isActive = false;
    }
    mouseEnterHandler() {
        this.isActive = true;
    }
    mouseLeaveHandler() {
        this.isActive = false;
    }
    render() {
        return (h(Host, null,
            h("span", { "aria-hidden": "true", class: this.unreadMessagesCount > 0
                    ? 'myporsche-unread-messages-indicator'
                    : 'myporsche-unread-messages-indicator hidden' }),
            h("p-button-pure", { class: "flyout-trigger", size: "inherit", icon: "user", active: this.isActive, onKeyDown: this.keyDownHandler, "aria-haspopup": "true", "aria-expanded": this.isActive ? 'true' : 'false', onClick: this.handleClick, id: this.screenReaderId, role: "button", "aria-label": "open myporsche flyout" },
                this.myPorscheText,
                h("p-text", { class: "login-status", color: "inherit", size: "inherit", ellipsis: "true", align: "left" }, this.loginStatus)),
            this.isActive ? (h("div", { class: "myporsche-items", "aria-labelledby": this.screenReaderId, role: "region" },
                h("phn-myporsche-items", { myPorscheCloseHandler: () => this.myPorscheCloseHandler() }))) : null));
    }
    static get is() { return "phn-myporsche-flyout"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["myporsche-flyout.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["myporsche-flyout.css"]
    }; }
    static get properties() { return {
        "myPorscheText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "my-porsche-text",
            "reflect": false
        },
        "loginStatus": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "login-status",
            "reflect": false
        },
        "env": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "env",
            "reflect": false,
            "defaultValue": "DEFAULT_ENVIRONMENT"
        },
        "unreadMessagesCount": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "unread-messages-count",
            "reflect": false,
            "defaultValue": "0"
        }
    }; }
    static get states() { return {
        "isActive": {}
    }; }
    static get listeners() { return [{
            "name": "mouseenter",
            "method": "mouseEnterHandler",
            "target": undefined,
            "capture": false,
            "passive": true
        }, {
            "name": "mouseleave",
            "method": "mouseLeaveHandler",
            "target": undefined,
            "capture": false,
            "passive": true
        }]; }
}
