// @porschehn/navigation/dist/collection/services/datalayer/custom-dimensions.js
export var ContextCollectionDimensions;
(function (ContextCollectionDimensions) {
    ContextCollectionDimensions["EventAction"] = "eventAction";
    ContextCollectionDimensions["ApplicationId"] = "applicationId";
    ContextCollectionDimensions["Language"] = "language";
    ContextCollectionDimensions["Country"] = "country";
    ContextCollectionDimensions["Timestamp"] = "timestamp";
    ContextCollectionDimensions["Environment"] = "environment";
})(ContextCollectionDimensions || (ContextCollectionDimensions = {}));
export var VisitorCollectionDimensions;
(function (VisitorCollectionDimensions) {
    VisitorCollectionDimensions["DeviceBrowserHeight"] = "deviceBrowserHeight";
    VisitorCollectionDimensions["DeviceBrowserWidth"] = "deviceBrowserWidth";
    VisitorCollectionDimensions["DeviceBrowserBreakpoint"] = "deviceBrowserBreakpoint";
    VisitorCollectionDimensions["DeviceBrowserOrientation"] = "deviceBrowserOrientation";
})(VisitorCollectionDimensions || (VisitorCollectionDimensions = {}));
export var PageExperienceCollectionDimensions;
(function (PageExperienceCollectionDimensions) {
    PageExperienceCollectionDimensions["PageId"] = "pageId";
    PageExperienceCollectionDimensions["PageName"] = "pageName";
})(PageExperienceCollectionDimensions || (PageExperienceCollectionDimensions = {}));
export var CarCollectionDimensions;
(function (CarCollectionDimensions) {
    CarCollectionDimensions["ModelRangeCode"] = "modelRangeCode";
    CarCollectionDimensions["ModelRangeName"] = "modelRangeName";
    CarCollectionDimensions["ModelSeriesCode"] = "modelSeriesCode";
    CarCollectionDimensions["ModelSeriesName"] = "modelSeriesName";
    CarCollectionDimensions["ModelName"] = "modelName";
    CarCollectionDimensions["Bodytype"] = "bodytype";
})(CarCollectionDimensions || (CarCollectionDimensions = {}));
export var ComponentClickCollectionDimension;
(function (ComponentClickCollectionDimension) {
    ComponentClickCollectionDimension["ClickElementType"] = "clickElementType";
    ComponentClickCollectionDimension["ClickElementId"] = "clickElementId";
    ComponentClickCollectionDimension["ClickElementName"] = "clickElementName";
    ComponentClickCollectionDimension["TargetUrl"] = "targetUrl";
})(ComponentClickCollectionDimension || (ComponentClickCollectionDimension = {}));
