// @porschehn/navigation/dist/collection/components/drawer-shops-and-more/drawer-shops-and-more.js
import { Component, Element, Listen, Prop, h } from '@stencil/core';
import StateTunnel from '../../services/state-tunnel';
import { EMPTY_ITEM } from '../../services/constants';
export class DrawerShopsAndMore {
    constructor() {
        this.item = EMPTY_ITEM;
        this.backButtonText = '';
        this.backButtonId = '';
        this.level = 2;
    }
    hideDrawer() {
        this.hide();
        this.closeActiveDrawer(this.level);
    }
    hide() {
        this.mainEl.style.display = 'none';
    }
    render() {
        this.hide();
        const buttonText = this.backButtonText ? this.backButtonText : this.menu && this.menu[0] ? this.menu[0].Text : '';
        const buttonId = this.backButtonId ? this.backButtonId : this.menu && this.menu[0] ? this.menu[0].PhraseId : '';
        const _headline = this.headline !== '' ? this.headline : this.item.Text;
        return (h("div", { class: `drawer-shops-and-more drawer-level-${this.level}-item-scroll drawer-level-${this.level}-item` },
            h("div", { class: "drawer-container" },
                h("div", { class: "drawer-back-and-close" },
                    h("phn-back-button", { class: "back-button", onClick: () => this.hideDrawer(), buttonId: buttonId, text: buttonText }),
                    h("phn-close-button", { class: "close-button" })),
                h("phn-headline", { text: _headline, class: "drawer-headline" }),
                this.item.Items &&
                    this.item.Items.map((item, index) => {
                        let innerOverlayEl;
                        return (h("div", null,
                            h("phn-list-item", { item: item, onClick: () => {
                                    if (!item.IsLinkItem) {
                                        innerOverlayEl.style.display = 'block';
                                        this.setActiveDrawer(innerOverlayEl, this.level + 1);
                                    }
                                }, class: 'shops-and-more-item shops-and-more-item-' + index }),
                            item.IsLinkItem ? null : (h("phn-drawer-shops-and-more", { headline: item.Items && item.Items.length === 1 ? item.Text : '', ref: (el) => (innerOverlayEl = el), item: item, level: this.level + 1, class: 'drawer-level-2 drawer-level-2-' + index, backButtonText: this.item.Text, backButtonId: this.item.PhraseId, closeActiveDrawer: this.closeActiveDrawer, setActiveDrawer: this.setActiveDrawer }))));
                    }))));
    }
    static get is() { return "phn-drawer-shops-and-more"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["drawer-shops-and-more.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["drawer-shops-and-more.css"]
    }; }
    static get properties() { return {
        "item": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "Item",
                "resolved": "Item",
                "references": {
                    "Item": {
                        "location": "import",
                        "path": "../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_ITEM"
        },
        "backButtonText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "back-button-text",
            "reflect": false,
            "defaultValue": "''"
        },
        "backButtonId": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "back-button-id",
            "reflect": false,
            "defaultValue": "''"
        },
        "headline": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "headline",
            "reflect": false
        },
        "level": {
            "type": "number",
            "mutable": false,
            "complexType": {
                "original": "number",
                "resolved": "number",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "level",
            "reflect": false,
            "defaultValue": "2"
        },
        "menu": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ContentItem[]",
                "resolved": "ContentItem[]",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../../services/state-tunnel"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "setActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(el: HTMLElement, level: number) => void",
                "resolved": "(el: HTMLElement, level: number) => void",
                "references": {
                    "HTMLElement": {
                        "location": "global"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "closeActiveDrawer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "(level: number) => void",
                "resolved": "(level: number) => void",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
    static get elementRef() { return "mainEl"; }
    static get listeners() { return [{
            "name": "closeEvent",
            "method": "hide",
            "target": undefined,
            "capture": false,
            "passive": false
        }]; }
}
StateTunnel.injectProps(DrawerShopsAndMore, ['menu']);
