// @porschehn/navigation/dist/collection/components/drawer-shops-and-more/drawer-shops-and-more.js
import { Component, Prop, h } from '@stencil/core';
import { EMPTY_ITEM } from '../../utility/constants';
import { navStateStore } from '../../state/nav-state-store';
export class DrawerShopsAndMore {
    constructor() {
        this.item = EMPTY_ITEM;
    }
    showDrawer(item, index) {
        const phnDrawerShopsAndMore = {
            backButton: this.item.Text,
            content: (h("phn-drawer-shops-and-more", { headline: item.Text, item: item, class: 'shops-and-more-drawer-level-2 drawer-level-2 drawer-level-2-' + index, slot: "content" }))
        };
        if (!item.Link && item.Items) {
            navStateStore.set('drawerContentStack', [...navStateStore.get('drawerContentStack'), phnDrawerShopsAndMore]);
        }
    }
    render() {
        return (h("div", { class: `drawer-shops-and-more` },
            h("div", { class: "drawer-container" },
                h("phn-headline", { text: this.headline, class: "drawer-headline" }),
                this.item.Items &&
                    this.item.Items.map((item, index) => {
                        return (h("div", null,
                            h("phn-list-item", { item: item, onClick: () => this.showDrawer(item, index), class: 'drawer-shops-and-more-item drawer-shops-and-more-item-' + index })));
                    }))));
    }
    static get is() { return "phn-drawer-shops-and-more"; }
    static get encapsulation() { return "scoped"; }
    static get originalStyleUrls() { return {
        "$": ["drawer-shops-and-more.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["drawer-shops-and-more.css"]
    }; }
    static get properties() { return {
        "item": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ContentItem",
                "resolved": "ContentItem",
                "references": {
                    "ContentItem": {
                        "location": "import",
                        "path": "../../entities/item"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "defaultValue": "EMPTY_ITEM"
        },
        "headline": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "headline",
            "reflect": false
        }
    }; }
}
